{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2fe0ab98-d0eb-4181-92f5-b0de69c6d1fe",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-09-30 19:11:12.535169: I tensorflow/core/util/port.cc:110] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.\n",
      "2025-09-30 19:11:12.539731: I tensorflow/tsl/cuda/cudart_stub.cc:28] Could not find cuda drivers on your machine, GPU will not be used.\n",
      "2025-09-30 19:11:12.657905: I tensorflow/tsl/cuda/cudart_stub.cc:28] Could not find cuda drivers on your machine, GPU will not be used.\n",
      "2025-09-30 19:11:12.660839: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
      "To enable the following instructions: AVX2 AVX512F AVX512_VNNI AVX512_BF16 AVX_VNNI AMX_TILE AMX_INT8 AMX_BF16 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2025-09-30 19:11:15.086878: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT\n"
     ]
    }
   ],
   "source": [
    "import skimage\n",
    "from skimage import io\n",
    "import cv2\n",
    "import matplotlib.pyplot as plt\n",
    "import tifffile as tfl\n",
    "import os\n",
    "import sys\n",
    "import numpy as np\n",
    "import scipy.io\n",
    "from skimage import img_as_float, img_as_uint\n",
    "from joblib import parallel, delayed\n",
    "from scipy.linalg import solve\n",
    "from skimage.util import view_as_blocks\n",
    "from scipy.signal import convolve2d , correlate2d, fftconvolve\n",
    "import shutil\n",
    "import glob\n",
    "from joblib import dump, load\n",
    "from skimage.morphology import disk\n",
    "from skimage.filters import median, difference_of_gaussians\n",
    "from skimage.registration import phase_cross_correlation as pcc\n",
    "from skimage.exposure import match_histograms\n",
    "from scipy.ndimage import shift\n",
    "import time\n",
    "import bardensr\n",
    "import pandas as pd\n",
    "from skimage.measure import label,regionprops_table\n",
    "from skimage import color\n",
    "from skimage.morphology import extrema\n",
    "from skimage.util import img_as_float\n",
    "from skimage.exposure import rescale_intensity\n",
    "from natsort import natsorted as nsort\n",
    "from skimage.transform import rescale\n",
    "from skimage.segmentation import expand_labels"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56ab1b5e-f62b-488b-b00c-d0bf10d6fc59",
   "metadata": {},
   "source": [
    "# I might want to change all sorted to nsort\n",
    "# I might also change all the basecall outputs to a folder keyed dictionary to be safe\n",
    "# Right now transform_coordinates method has cell pos and hyb pos inverted to make it correct, that means I have to interchange xy in the basecalls for hyb and cellpose label saving to revert to proper function transform_coord\n",
    "# Also check for any unintentional aliasing, if possible change all direct references to .copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5414a90a-5dfa-4a67-b118-bf5a2133b32e",
   "metadata": {},
   "source": [
    "- To install fiji: wget https://downloads.imagej.net/fiji/archive/20191027-2045/fiji-linux64.zip, unzip fiji-linux64.zip\n",
    "- To install MIST: Navigate to FIJI.app/plugins and run wget https://sites.imagej.net/NIST-ISG-MIST/plugins/MIST_.jar-20231103172358\n",
    "    - rename to MIST_.jar\n",
    "- To install fftw follow https://github-wiki-see.page/m/usnistgov/MIST/wiki/Install-Guide\n",
    "    - copy libfftw3.so and libfftw3f.so to your FIJI.app/lib/fftw folder\n",
    "- You might need to copy some jar files from windows installation of FIJI.app/jar to same place on ubuntu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "23ab8c74-a41b-4dfe-b98c-69ebfa941dac",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Operating in headless mode - the original ImageJ will have limited functionality.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initialization successful!\n",
      "ImageJ version: 2.16.0/1.54p\n",
      "Legacy service is active: True\n"
     ]
    }
   ],
   "source": [
    "import imagej\n",
    "import scyjava\n",
    "\"\"\"\n",
    "This cell is to initialize FIJI\n",
    "\"\"\"\n",
    "\n",
    "scyjava.config.add_option('-Xmx6g')\n",
    "\n",
    "pth_fiji = r'/home/nrg/Fiji.app' \n",
    "\n",
    "try:\n",
    "    \n",
    "    ij = imagej.init(pth_fiji, add_legacy=True)\n",
    "    \n",
    "    print(\"Initialization successful!\")\n",
    "    print(f\"ImageJ version: {ij.getVersion()}\")\n",
    "    \n",
    "    print(f\"Legacy service is active: {ij.legacy.isActive()}\")\n",
    "\n",
    "except Exception as e:\n",
    "    print(f\" An error occurred during initialization: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2670e563-1c01-4237-a3a9-50ff3aae31a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def minmax_norm(x):\n",
    "    \"\"\"\n",
    "    Preprocessing function: \n",
    "    Image normalization by its minimum and maximum\n",
    "    \"\"\"\n",
    "    y=np.divide((x-np.min(x,axis=None)),np.max(x,axis=None)-np.min(x,axis=None))\n",
    "    return y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "81ee34ff-7f58-43aa-b9e4-6608abdb12be",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def uint16m(x):\n",
    "    \"\"\"\n",
    "    Preprocessing function:\n",
    "    MATLAB equivalent of uint16-does clipping between data bounds and then converts\n",
    "    \"\"\"\n",
    "    y=np.uint16(np.clip(np.round(x),0,65535))\n",
    "    return y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4e553d3a-603f-4a68-be9c-5f818e5a8f59",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_folders(pth):\n",
    "    \"\"\"\n",
    "    Universal function:\n",
    "    Scans the processed directory to get total tiles (folders),names, positions and x and y grid points \n",
    "    \"\"\"\n",
    "    dr=nsort(glob.glob(os.path.join(pth,'processed','MAX*')))\n",
    "    folder=[]\n",
    "    xp=[]\n",
    "    yp=[]\n",
    "    pos=[]\n",
    "    for i,j in enumerate(dr):\n",
    "        folder.append(str(j).split('/')[-1])\n",
    "        temp=str(j).split('/')[-1].split('_')\n",
    "        pos.append(temp[1])\n",
    "        xp.append(int(temp[2])+1)\n",
    "        yp.append(int(temp[3])+1)\n",
    "    return folder,pos,xp,yp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "29daa587-1a98-428a-93da-f8d236b37a82",
   "metadata": {},
   "outputs": [],
   "source": [
    "def denoise_n2v_call(pth,basedir,fname,config_file='cfg.yaml'):\n",
    "    run_diff_env_scripts(\"denoising_n2v\",config_file,pth=pth,basedir=basedir,fname=fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "245e5191-9fee-46fb-9cce-76f86e2565d1",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def back_sub_opencv_open(I,radius,pth,name,num_c,writefile):\n",
    "    \"\"\"\n",
    "    Preprocessing function:\n",
    "    Background subtraction per tile using opencv morphological opening\n",
    "    \"\"\"\n",
    "    I=I.copy()\n",
    "    I_filtered=np.zeros_like(I)\n",
    "    I_rem=I[num_c:,:,:]\n",
    "    I=I[0:num_c,:,:]\n",
    "    k=cv2.getStructuringElement(cv2.MORPH_ELLIPSE,(radius,radius))\n",
    "    for i in range(len(I)):\n",
    "        bck=cv2.morphologyEx(I[i,:,:], cv2.MORPH_OPEN, kernel= k)\n",
    "        I_filtered[i,:,:]=I[i,:,:]-np.expand_dims(bck,0)\n",
    "    \n",
    "    I_filtered[num_c:,:,:]=I_rem    \n",
    "    I_filtered=uint16m(I_filtered)\n",
    "    if writefile:\n",
    "        tfl.imwrite(os.path.join(pth,name),I_filtered,photometric='minisblack')\n",
    "    return I_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f5783f27-1ca5-4dba-9090-51cb939ab733",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def channel_alignment(I_filtered,fname,config_pth,pth,name,num_c,is_affine,writefile):\n",
    "    \"\"\"\n",
    "    Preprocessing function:\n",
    "    Channel alignment per tile using only translations \n",
    "    \"\"\"\n",
    "    I_filtered=I_filtered.copy()\n",
    "    Ishifted=np.zeros_like(I_filtered)\n",
    "    I_rem=I_filtered[num_c:,:,:]\n",
    "    I_filtered=I_filtered[0:num_c,:,:]\n",
    "    chshift=scipy.io.loadmat(os.path.join(config_pth,fname))['chshift20x']\n",
    "    for i in range(chshift.shape[0]):\n",
    "        if is_affine:\n",
    "            tform=chshift[i] # refine this later on-ng\n",
    "        else:\n",
    "            tform=skimage.transform.SimilarityTransform(translation=-chshift[i,:]) # remember this takes -shifts or use tform.inverse-ng\n",
    "            \n",
    "        It=skimage.transform.warp(np.squeeze(I_filtered[i,:,:]),tform,preserve_range=True,output_shape=(I_filtered.shape[1],I_filtered.shape[2]))\n",
    "        Ishifted[i,:,:]=np.expand_dims(It,0)\n",
    "    Ishifted[num_c:,:,:]=I_rem    \n",
    "    Ishifted=uint16m(Ishifted)\n",
    "    if writefile:\n",
    "        tfl.imwrite(os.path.join(pth,name),Ishifted,photometric='minisblack')\n",
    "    return Ishifted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "99543ec4-8ba6-49ce-8312-4a37358c95c6",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def bleedthrough_linear(Is,num_c,config_pth,fname,pth,name,writefile):\n",
    "    \"\"\"\n",
    "    Preprocessing function:\n",
    "    Bleedthrough correction per tile \n",
    "    \"\"\"\n",
    "    Is=Is.copy()\n",
    "    Icorrected=np.zeros_like(Is)\n",
    "    chprofile=scipy.io.loadmat(os.path.join(config_pth,fname))['chprofile20x']\n",
    "    Ishifted2=np.float64(Is[0:num_c,:,:])\n",
    "    I_rem=Is[num_c:,:,:]\n",
    "    Icorrected=(np.reshape((np.linalg.solve(np.transpose(chprofile),((np.reshape(Ishifted2,(num_c,-1),order='F'))))),(num_c,Ishifted2.shape[1],Ishifted2.shape[2]),order='F'))\n",
    "    Icorrected=uint16m(Icorrected)\n",
    "    Icorrected=np.append(Icorrected,I_rem,axis=0)\n",
    "    if writefile:\n",
    "        tfl.imwrite(os.path.join(pth,name),Icorrected,photometric='minisblack')\n",
    "    return Icorrected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "99861fb5-2194-4200-b488-d3381ddc72ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "def organize_files(pth='D:/HTNA24_new/',cyclename='geneseq'):\n",
    "    \"\"\"\n",
    "    Preprocessing function:\n",
    "    Copy all cycle geneseq files from experiment directory to each folder representing one tile\n",
    "    \"\"\"\n",
    "    print('Copying files in the correct folders...')\n",
    "    genefolders=sorted(glob.glob(os.path.join(pth,cyclename+\"*\")))\n",
    "    #print(genefolders)\n",
    "    for f,i in enumerate(genefolders):\n",
    "        if not(os.path.isdir(f)):\n",
    "            f=[]\n",
    "    \n",
    "    try:\n",
    "        os.makedirs(os.path.join(pth,'processed'),exist_ok=True)\n",
    "        #print(pthw)\n",
    "    except Exception:\n",
    "        print('Directory already exists\\n')\n",
    "    for i,f2 in enumerate(genefolders):\n",
    "        print(f'Organizing folder {genefolders[i]}\\n')\n",
    "        files=sorted(glob.glob(os.path.join(pth,genefolders[i],'MAX*.tif')))\n",
    "        for j,f1 in enumerate(files):\n",
    "            name=f1.split('MAX')\n",
    "            fname=name[1].split('.')\n",
    "            os.makedirs(os.path.join(pth,'processed','MAX'+fname[0]),exist_ok=True)\n",
    "            shutil.copy(f1,os.path.join(pth,'processed','MAX'+fname[0],cyclename+f2.partition(cyclename)[-1]+'.tif'))\n",
    "            print(f'Copied {j}')\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8c54c746-d761-437b-8552-492e22d7747e",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def geneseq_cycle_alignment_block_correlation(imagename,templatename,subsample_rate,do_coarse,resize_factor,block_size,num_c,numcr,pth,folder,name):\n",
    "    \"\"\"\n",
    "    Preprocessing function:\n",
    "    Given two images-two cycles of geneseq-align image (moving) to template (fixed) using windowed phase correlation based registration\n",
    "    Writes aligned image, returns aligned image and transformation matrix\n",
    "    \"\"\"\n",
    "    os.makedirs(os.path.join(pth,folder),exist_ok=True)\n",
    "        \n",
    "    moving=tfl.imread(os.path.join(pth,imagename),key=range(0,num_c,1))\n",
    "    fixed=tfl.imread(os.path.join(pth,templatename),key=range(0,num_c,1))\n",
    "    moving_sum=np.double(np.sum(moving,axis=0))\n",
    "    moving_sum=np.divide(moving_sum,np.max(moving_sum,axis=None))\n",
    "    fixed_sum=np.double(np.sum(fixed,axis=0))\n",
    "    fixed_sum=np.divide(fixed_sum,np.max(fixed_sum,axis=None))\n",
    "    sz=fixed_sum.shape\n",
    "    b_x=np.floor(sz[0]/block_size)\n",
    "    b_y=np.floor(sz[1]/block_size)\n",
    "\n",
    "\n",
    "    if b_x*block_size!=sz[0]:\n",
    "        fixed_sum=fixed_sum[0:b_x*block_size-1,:]\n",
    "        moving_sum=moving_sum[0:b_x*block_size-1,:]\n",
    "\n",
    "    if b_y*block_size!=sz[1]:\n",
    "        fixed_sum=fixed_sum[:,0:b_y*block_size-1]\n",
    "        moving_sum=moving_sum[:,0:b_y*block_size-1]\n",
    "        \n",
    "    moving_rescaled=np.uint8(skimage.transform.rescale(moving_sum,resize_factor)*255) # check if this uint8 needs to be changed as per matlab standard-ng\n",
    "    fixed_rescaled=np.uint8(skimage.transform.rescale(fixed_sum,resize_factor)*255)\n",
    "    moving_split=view_as_blocks(moving_rescaled, block_shape=(block_size*resize_factor,block_size*resize_factor))\n",
    "    fixed_split=view_as_blocks(fixed_rescaled, block_shape=(block_size*resize_factor,block_size*resize_factor))\n",
    "    fixed_split_lin=np.reshape(fixed_split,(-1,fixed_split.shape[2],fixed_split.shape[3]))\n",
    "    \n",
    "    fixed_split_sum=[np.sum(j,axis=None) for i,j in enumerate(fixed_split_lin)]\n",
    "    idx=np.argsort(fixed_split_sum)[::-1]\n",
    "    fixed_split_sum=np.reshape(fixed_split_sum,(fixed_split.shape[0],fixed_split.shape[1]))\n",
    "   \n",
    "    moving_split_lin=np.reshape(moving_split,(-1,moving_split.shape[2],moving_split.shape[3]))\n",
    "    \n",
    "    xcorr2=lambda a,b: fftconvolve(a,np.rot90(b,k=2))\n",
    "    c=np.zeros((fixed_split_lin.shape[1]*2-1,fixed_split_lin.shape[2]*2-1))\n",
    "    for i in range(np.int32(np.round(fixed_split_lin.shape[0]/subsample_rate))): # check for int32-ng\n",
    "        if np.max(fixed_split_lin[idx[i]],axis=None)>0:\n",
    "            c=c+xcorr2(np.double(fixed_split_lin[idx[i]]),np.double(moving_split_lin[idx[i]]))\n",
    "\n",
    "    shift_yx=np.unravel_index(np.argmax(c),c.shape)\n",
    "\n",
    "    yoffset=-np.array([(shift_yx[0]+1-fixed_split_lin.shape[1])/resize_factor])\n",
    "    xoffset=-np.array([(shift_yx[1]+1-fixed_split_lin.shape[2])/resize_factor])\n",
    "\n",
    "    idx_minxy=np.argmin(np.abs(xoffset)+np.abs(yoffset))\n",
    "\n",
    "    tform=skimage.transform.SimilarityTransform(translation=[xoffset[idx_minxy], yoffset[idx_minxy]])\n",
    "\n",
    "    moving_aligned=np.zeros_like(moving)\n",
    "    for i in range(moving.shape[0]):\n",
    "        moving_aligned[i,:,:]=np.expand_dims(skimage.transform.warp((np.squeeze(moving[i,:,:])),tform,preserve_range=True),0)#,output_shape=(moving.shape[1],moving.shape[2])),0)# check if output size specification is necessary -ng\n",
    "\n",
    "    moving_aligned=uint16m(moving_aligned)\n",
    "    moving_aligned_full=moving_aligned.copy()\n",
    "    if numcr>num_c:\n",
    "        moving_last_frame=tfl.imread(os.path.join(pth,imagename),key=range(num_c,numcr,1))\n",
    "        if moving_last_frame.ndim==2:\n",
    "            moving_last_frame=np.expand_dims(moving_last_frame,axis=0)\n",
    "        moving_aligned_full=np.append(moving_aligned_full,moving_last_frame,axis=0)\n",
    "        \n",
    "    tfl.imwrite(os.path.join(pth,folder,name),moving_aligned_full,photometric='minisblack')\n",
    "    return moving_aligned, tform"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "bad51d2d-0ce6-4d6c-ac86-11b5f9c8d117",
   "metadata": {},
   "outputs": [],
   "source": [
    "# have not included default zeros shiftmat on non-existent input--ng\n",
    "def mmseqalignmentsinglethread_local_subsample(pth,config_pth,cyclename,num_initial_c,num_later_c,resize_factor,block_size,chshift_filename,chprofile_filename,radius,is_affine=0,num_c=4,subsample_rate=1,do_coarse=0):\n",
    "    \"\"\"\n",
    "    Preprocessing function:\n",
    "    Does background subtraction, channel alignment, bleedthrough correction and cycle alignement for sequencing images-operates on one tile\n",
    "    \"\"\"\n",
    "    cycle_im=sorted(glob.glob(os.path.join(pth,cyclename+\"*\")))\n",
    "    os.makedirs(os.path.join(pth,'original'),exist_ok=True)\n",
    "    os.makedirs(os.path.join(pth,'aligned'),exist_ok=True)\n",
    "    fixed_filename=cycle_im.copy()\n",
    "    gtforms=[]\n",
    "\n",
    "    for i in range(len(cycle_im)):\n",
    "\n",
    "        filename=fixed_filename[i].split(\"/\")[-1]\n",
    "        fixed_filename[i]='fixed'+filename\n",
    "        if i==0:\n",
    "            num_cr=num_initial_c\n",
    "            templatename=fixed_filename[i]\n",
    "        else:\n",
    "            num_cr=num_later_c\n",
    "\n",
    "        I=tfl.imread(cycle_im[i],key=range(0,num_cr,1))\n",
    "\n",
    "        if radius>0:\n",
    "            Ibcksub=back_sub_opencv_open(I,radius,pth,filename,num_c,0) #BACKGROUND SUBTRACTION-NG\n",
    "        else:\n",
    "            Ibcksub=I\n",
    "    \n",
    "        Ibcksub_shifted=channel_alignment(Ibcksub,chshift_filename,config_pth,pth,'bck_sub'+filename,num_c,is_affine,0) # last argument is to not to write intermediate file-ng CHANNEL ALIGN-NG\n",
    "        Ibcksub_shifted_btcorr=bleedthrough_linear(Ibcksub_shifted,num_c,config_pth,chprofile_filename,pth,fixed_filename[i],1) # BLEEDTHROUGH CORRECTION-NG\n",
    "        _,gtforms_ind=geneseq_cycle_alignment_block_correlation(fixed_filename[i],templatename,subsample_rate,do_coarse,resize_factor,block_size,num_c,num_cr,pth,'aligned','aligned'+filename)\n",
    "        gtforms.append(gtforms_ind)\n",
    "\n",
    "    dump(gtforms,os.path.join(pth,'tforms'+cyclename+'.joblib'))\n",
    "    [os.remove(f) for f in glob.glob(os.path.join(pth,'fixed*.tif'))]\n",
    "    [shutil.move(f1,os.path.join(pth,'original/'+f1.split('/')[-1])) for f1 in glob.glob(os.path.join(pth,cyclename+'*.tif'))]\n",
    "    print(f'BACKGROUND SUBTRACTION, CHANNEL ALIGNEMENT, BLEEDTHROUGH CORRECTION AND CYCLE ALIGNEMENT COMPLETE FOR FOLDER {pth.split(\"/\")[-1]}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "692024d2-c7bc-43a3-ae4f-5f66f96f1a98",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rgbout(pthb,config_pth,folders,cyclename,shiftmat_name=None,profilemat_name=None,align=0,bleed_correct=0,bcksub=0,m=0,strength=0.6,radius=0,num_c=4,is_affine=0):\n",
    "    \"\"\"\n",
    "    Preprocessing function:\n",
    "    From the aligned images-prepares an RGB image-channels to colors--this will be stitched later \n",
    "    \"\"\"\n",
    "    pthl=os.path.join(pthb,'processed',folders)\n",
    "    os.makedirs(os.path.join(pthl,'RGB'),exist_ok=True)\n",
    "    gname='aligned'+cyclename\n",
    "    aligned_im=sorted(glob.glob(os.path.join(pthl,'aligned',gname+\"*.tif\")))\n",
    "    filename=aligned_im.copy()\n",
    "\n",
    "    for i in range(len(aligned_im)):\n",
    "        I=tfl.imread(aligned_im[i],key=range(0,4,1))\n",
    "        Ifilt=np.zeros_like(I)\n",
    "        I_rgb=np.zeros((3,I.shape[1],I.shape[2]),dtype=np.double)\n",
    "        if bcksub:\n",
    "            Ibck=back_sub_opencv_open(I,radius,aligned_im[i],'bcksub.tif',num_c,0)\n",
    "        else:\n",
    "            Ibck=I\n",
    "        if align:\n",
    "            Ishifted=channel_alignment(Ibck,shiftmat_name,config_pth,aligned_im[i],'aligned.tif',num_c,is_affine,0)\n",
    "        else:\n",
    "            Ishifted=I\n",
    "        if bleed_correct:\n",
    "            Ifin=bleedthrough_linear(Ishifted,num_c,config_pth,profilemat_name,pth,aligned_im[i],'bcksub_aligned_btcorr.tif',0)\n",
    "        else:\n",
    "            Ifin=I\n",
    "        if m>0:\n",
    "            for ch in range(Ifin.shape[0]):\n",
    "                Ifilt[ch,:,:]=median(Ifin[ch,:,:],disk(m))\n",
    "        else:\n",
    "            Ifilt=Ifin\n",
    "            \n",
    "        I_rgb[0,:,:]=np.double(Ifilt[1,:,:]+Ifilt[2,:,:]+Ifilt[3,:,:])*strength\n",
    "        I_rgb[1,:,:]=np.double(Ifilt[1,:,:]+Ifilt[0,:,:]+Ifilt[3,:,:])*strength\n",
    "        I_rgb[2,:,:]=np.double(Ifilt[0,:,:]+Ifilt[2,:,:]+Ifilt[3,:,:])*strength\n",
    "\n",
    "        Irgb=np.uint8(np.clip(I_rgb,0,255))\n",
    "        tfl.imwrite(os.path.join(pthl,'RGB','RGB'+filename[i].split('/')[-1]),np.transpose(Irgb,axes=(1,2,0)),photometric='rgb')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ef674762-c6a6-44f4-80ba-1bf164a34ead",
   "metadata": {},
   "outputs": [],
   "source": [
    "def align_hyb_to_gene(moving_hyb,template_sum_norm,Ibcksub_shifted_btcorr):\n",
    "    \"\"\"\n",
    "    Preprocessing function:\n",
    "    Given two images from one tile--hyb and gene--aligns hyb to geneseq\n",
    "    Returns aligned image and transformation matrix\n",
    "    \"\"\"\n",
    "    moving=moving_hyb.copy()\n",
    "    moving=np.squeeze(moving,axis=0)\n",
    "    moving_norm=np.divide(moving,np.max(moving,axis=None))\n",
    "    moving=moving_norm\n",
    "    fixed=template_sum_norm.copy()\n",
    "    moving=np.uint8(np.clip(moving*255,0,255))\n",
    "    fixed=np.uint8(np.clip(fixed*255,0,255))\n",
    "\n",
    "    hmatched_moving = match_histograms(moving, fixed)\n",
    "    \n",
    "    shift_values,_,_ = pcc(fixed, hmatched_moving, upsample_factor=100)\n",
    "    \n",
    "    tform_hyb=skimage.transform.SimilarityTransform(translation=(-shift_values[1],-shift_values[0]))\n",
    "\n",
    "    Ih_aligned=np.zeros_like(Ibcksub_shifted_btcorr)\n",
    "    for i in range(Ibcksub_shifted_btcorr.shape[0]):\n",
    "        Ih_aligned[i,:,:]=skimage.transform.warp(np.squeeze(Ibcksub_shifted_btcorr[i,:,:]),tform_hyb,preserve_range=True,output_shape=(template_sum_norm.shape[0],template_sum_norm.shape[1]))\n",
    "    \n",
    "    return Ih_aligned,tform_hyb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2cb09196-7e2e-4f5e-9bf8-8f0f99a0e1c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_hyb_images(pth,config_pth,chshift_hyb_filename,chprofile_hyb_filename,cycle_name='n2vhyb',ref_cycle_name='alignedn2vgene',reg_cycle=1,dapi_shift=[],dapi_ch_num=5,num_c_hyb=6,num_c=4,is_affine=0,radius=31,ch_i=2):\n",
    "    \"\"\"\n",
    "    Preprocessing function:\n",
    "    Does background subtraction, channel alignment, bleedthrough correction and alignement to geneseq for hyb images-operates on one tile\n",
    "    \"\"\"\n",
    "        \n",
    "    os.makedirs(os.path.join(pth,'original'),exist_ok=True)\n",
    "    os.makedirs(os.path.join(pth,'aligned'),exist_ok=True)\n",
    "    \n",
    "    cycle_im=sorted(glob.glob(os.path.join(pth,cycle_name+\"*\")))\n",
    "    ref_cycle_im=sorted(glob.glob(os.path.join(pth,'aligned',ref_cycle_name+\"*\")))\n",
    "    filename=cycle_im.copy()\n",
    "    ch=ch_i*np.ones((1,len(cycle_im)),dtype=np.int32)\n",
    "    \n",
    "    template=tfl.imread(ref_cycle_im[reg_cycle],key=range(0,num_c,1))\n",
    "    \n",
    "    template_sum=np.double(np.sum(template,axis=0))\n",
    "    template_sum_norm=np.divide(template_sum,np.max(template_sum,axis=None))\n",
    "    \n",
    "    \n",
    "    for i in range(len(cycle_im)):\n",
    "        filename[i]=filename[i].split('/')[-1]\n",
    "        Ih=tfl.imread(cycle_im[i],key=range(0,num_c_hyb,1))\n",
    "        if radius>0:\n",
    "            Ibcksub=back_sub_opencv_open(Ih,radius,pth,'bcksub'+filename[i],num_c_hyb-1,0)\n",
    "        else:\n",
    "            Ibcksub=Ih\n",
    "            \n",
    "        Ibcksub_shifted=channel_alignment(Ibcksub,chshift_hyb_filename,config_pth,pth,'shifted'+filename[i],num_c,is_affine,0)\n",
    "        Ibcksub_shifted_btcorr=bleedthrough_linear(Ibcksub_shifted,num_c,config_pth,chprofile_hyb_filename,pth,'fixed'+filename[i],1) # BLEEDTHROUGH CORRECTION-NG\n",
    "        moving_hyb=Ibcksub_shifted_btcorr[ch[i],:,:]\n",
    "        [Ihyb_aligned,tform_hyb]=align_hyb_to_gene(moving_hyb,template_sum_norm,Ibcksub_shifted_btcorr)\n",
    "        tfl.imwrite(os.path.join(pth,'aligned','aligned'+filename[i]),Ihyb_aligned,photometric='minisblack')\n",
    "        tfl.imwrite(os.path.join(pth,'aligned','alignednuclear.tif'),Ihyb_aligned,photometric='minisblack')\n",
    "        dump(tform_hyb,os.path.join(pth,'tforms_hyb_to_gene.joblib'))\n",
    "        [os.remove(f) for f in glob.glob(os.path.join(pth,'fixed*.tif'))]\n",
    "        [shutil.move(f1,os.path.join(pth,'original/'+f1.split('/')[-1])) for f1 in glob.glob(os.path.join(pth,cycle_name+'*.tif'))]\n",
    "        print(f'BACKGROUND SUBTRACTION, CHANNEL ALIGNEMENT, BLEEDTHROUGH CORRECTION AND HYB TO GENE ALIGNEMENT COMPLETE FOR FOLDER {cycle_im[i].split(\"/\")[-2]}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a012a510-ae25-468f-800d-ef68aab13b2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def single_align_bc_to_gene(pth,bcfname='bcseq', genefname='gene',bcidx=4,geneidx=4,ch_bc=4,ch_bc_init=5):\n",
    "    \"\"\"\n",
    "    Preprocessing function:\n",
    "    Aligns bcseq0 to geneseq0, reference channels-brightfield at this point, using phase correlation based registration, writes merged file in ??? [NEED TO CHECK WHERE IS COMP FILE WRITTEN]-operates on one tile\n",
    "    Writes aligned file\n",
    "    \"\"\"\n",
    "   \n",
    "    bcfiles=sorted(glob.glob(os.path.join(pth,'*'+bcfname+'*.tif')))\n",
    "    genefiles=sorted(glob.glob(os.path.join(pth,'*'+genefname+'*.tif')))\n",
    "    filename=bcfiles.copy()\n",
    "    \n",
    "    geneim=tfl.imread(genefiles[0],key=geneidx)\n",
    "    bcim=tfl.imread(bcfiles[0],key=bcidx)\n",
    "   \n",
    "    \n",
    "    moving=bcim.copy()\n",
    "    moving= difference_of_gaussians(moving, 2, 40) \n",
    "    moving_norm=np.divide(moving,np.max(moving,axis=None))\n",
    "    moving=moving_norm.copy()\n",
    "    fixed=geneim.copy()\n",
    "    fixed= difference_of_gaussians(fixed, 2, 40) \n",
    "    fixed_norm=np.divide(fixed,np.max(fixed,axis=None))\n",
    "    fixed=fixed_norm.copy()\n",
    "    moving=np.uint8(np.clip(moving*255,0,255))\n",
    "    fixed=np.uint8(np.clip(fixed*255,0,255))\n",
    "    \n",
    "    hmatched_moving = match_histograms(moving, fixed)\n",
    "    \n",
    "    shift_values,_,_ = pcc(fixed, hmatched_moving, upsample_factor=100)\n",
    "    \n",
    "    tform_bc=skimage.transform.SimilarityTransform(translation=(-shift_values[1],-shift_values[0]))\n",
    "    aligned=np.zeros((3,fixed.shape[0],fixed.shape[1]),dtype=np.double)\n",
    "    # aligned[0,:,:]=uint16m(geneim)\n",
    "    # aligned[1,:,:]=uint16m(skimage.transform.warp(bcim,tform_bc,preserve_range=True,output_shape=(fixed.shape[0],fixed.shape[1])))\n",
    "    # tfl.imwrite(os.path.join(pth,'comp.tif'),aligned,photometric='minisblack')\n",
    "    aligned[0,:,:]=np.double(fixed)*0.5\n",
    "    aligned[1,:,:]=np.double(skimage.transform.warp(moving,tform_bc,preserve_range=True,output_shape=(fixed.shape[0],fixed.shape[1])))*0.5\n",
    "     \n",
    "\n",
    "    aligned=np.uint8(np.clip(aligned,0,255))\n",
    "    tfl.imwrite(os.path.join(pth,'comp.tif'),np.transpose(aligned,axes=(1,2,0)),photometric='rgb')\n",
    "\n",
    "\n",
    "\n",
    "    for j in range(len(bcfiles)):\n",
    "        if j==0:\n",
    "            ch_r=ch_bc_init\n",
    "        else:\n",
    "            ch_r=ch_bc\n",
    "        Ibc=tfl.imread(bcfiles[j],key=range(0,ch_r,1))\n",
    "        Ibc_aligned=np.zeros_like(Ibc)\n",
    "        for i in range(Ibc.shape[0]):\n",
    "            Ibc_aligned[i,:,:]=skimage.transform.warp(np.squeeze(Ibc[i,:,:]),tform_bc,preserve_range=True,output_shape=(fixed.shape[0],fixed.shape[1]))\n",
    "        tfl.imwrite(os.path.join(pth,'reg'+filename[j].split('/')[-1]),Ibc_aligned,photometric='minisblack')\n",
    "    \n",
    "    return tform_bc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4214e4dc-f3e8-4476-8ee7-2588f48f9dab",
   "metadata": {},
   "outputs": [],
   "source": [
    "def align_bc_to_gene(pthb,bcfname='n2vbcseq', genefname='n2vgene',bcidx=4,geneidx=4,ch_bc=4,ch_bc_init=5):\n",
    "    \"\"\"\n",
    "    Preprocessing function:\n",
    "    calls single_slign_bc_to_gene to align bcseq0 to geneseq0 for all tiles\n",
    "    writes transformation matrix per tile\n",
    "    \"\"\"\n",
    "    os.makedirs(os.path.join(pthb,'processed','BCtoGeneAlignementComparison'),exist_ok=True)\n",
    "    # ALSO, I AM ONLY ALIGNING RAW FILES HERE-NOT N2V-DOESN'T MATTER BECAUSE BF IS NOT CHANGED-WHAT HAPPENS WHEN WE SWITCH TO DAPI IS ANOTHER STORY\n",
    "    #organize_files(pthb,cyclename='bcseq')\n",
    "    folders,_,_,_=get_folders(pthb)\n",
    "    for i in range(len(folders)):\n",
    "        pthl=os.path.join(pthb,'processed',folders[i])\n",
    "        [shutil.move(f1,os.path.join(pthl,f1.split('/')[-1])) for f1 in glob.glob(os.path.join(pthl,'original','n2v*.tif'))]\n",
    "        [os.remove(f) for f in glob.glob(os.path.join(pthl,'*reg*.tif'))]\n",
    "        tform_bc=single_align_bc_to_gene(pthl,bcfname,genefname,bcidx,geneidx,ch_bc,ch_bc_init)\n",
    "        dump(tform_bc,os.path.join(pthl,'tforms_bc_to_gene.joblib'))\n",
    "        [shutil.move(f1,os.path.join(pthb,'processed','BCtoGeneAlignementComparison',bcfname+'2'+genefname+'_'+folders[i]+'.tif')) for f1 in glob.glob(os.path.join(pthl,'comp.tif'))]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "689b4119-5063-41ae-ae5c-a332492052de",
   "metadata": {},
   "outputs": [],
   "source": [
    "def register_seq_images_subsample(pthb,config_pth,chshift_filename,chprofile_filename,cyclename,is_affine=0,num_initial_c=5, num_later_c=4,num_c=4,subsample_rate=4,do_coarse=0,resize_factor=2, block_size=128,radius=31,local=1,strength=0.6,m=0):\n",
    "    \"\"\"\n",
    "    Preprocessing function:\n",
    "    For all the tiles-calls functions for background subtraction, bleed-through correction, channel alignement and cycles alignement-creates RGB--moves n2v files back to original subfolder within each folder for tile\n",
    "    \"\"\"\n",
    "    folders,_,_,_=get_folders(pthb)\n",
    "    for i in range(len(folders)):\n",
    "        pthl=os.path.join(pthb,'processed',folders[i])\n",
    "        if os.path.isdir(os.path.join(pthl,'original')):\n",
    "            [shutil.move(f1,os.path.join(pthl,f1.split('/')[-1])) for f1 in glob.glob(os.path.join(pthl,'original/','*.tif'))]\n",
    "        if local==1:\n",
    "            mmseqalignmentsinglethread_local_subsample(pthl,config_pth,cyclename,num_initial_c,num_later_c,resize_factor,block_size,chshift_filename,chprofile_filename,radius,is_affine=0,num_c=4,subsample_rate=1,do_coarse=0)\n",
    "        else:\n",
    "            print('code some other function here')\n",
    "        rgbout(pthb,config_pth,folders[i],cyclename,shiftmat_name=None,profilemat_name=None,align=0,bleed_correct=0,bcksub=0,m=0,strength=strength,radius=0,num_c=num_c,is_affine=is_affine)\n",
    "\n",
    "        [shutil.move(f1,os.path.join(pthl,'original',f1.split('/')[-1])) for f1 in glob.glob(os.path.join(pthl,'n2v*.tif'))]\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "bee0c4a1-42a4-4d3a-9924-9cd052e359eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_bc_cycle(pthb,config_pth,basedir,chshift_filename,chprofile_filename,config_file='cfg.yaml',is_affine=0,num_initial_c=5,num_later_c=4,num_c=4,subsample_rate=4,do_coarse=0,resize_factor=2,block_size=128,radius=31,local=1,cyclename='regn2vbc',bcidx=4,geneidx=4,ch_bc=4,ch_bc_init=5,bcfname='n2vbcseq',genefname='n2vgene',strength=0.6,m=0):\n",
    "    \"\"\"\n",
    "    Preprocessing function:\n",
    "    For bcseq--calls functions that align bcseq0 to geneseq0, preprocess bcseq--for all tiles\n",
    "    \"\"\"   \n",
    "    t0=time.time()\n",
    "    #organize_files(pthb,cyclename='bcseq')\n",
    "    #denoise_n2v_call(pthb,basedir,fname='bcseq',config_file=config_file)\n",
    "    #align_bc_to_gene(pthb,bcfname, genefname,bcidx,geneidx,ch_bc,ch_bc_init)\n",
    "    \n",
    "    register_seq_images_subsample(pthb,config_pth,chshift_filename,chprofile_filename,cyclename,is_affine,num_initial_c, num_later_c,num_c,subsample_rate,do_coarse,resize_factor, block_size,radius,local,strength,m)\n",
    "    \n",
    "    tn=time.time()\n",
    "    \n",
    "    print(f\"Total time {tn-t0} seconds\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "741b4718-c3c9-424f-9acc-cdd42f667d79",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_geneseq_cycle(pthb,config_pth,basedir,chshift_filename,chprofile_filename,config_file='cfg.yaml',is_affine=0,num_initial_c=5,num_later_c=4,num_c=4,subsample_rate=4,do_coarse=0,resize_factor=2,block_size=128,radius=31,local=1,cyclename='n2vgene',strength=0.6,m=0):\n",
    "    \"\"\"\n",
    "    Preprocessing function:\n",
    "    For bcseq--calls functions that align bcseq0 to geneseq0, preprocess bcseq--for all tiles\n",
    "    \"\"\"   \n",
    "    t0=time.time()\n",
    "    organize_files(pthb,cyclename='geneseq')\n",
    "    denoise_n2v_call(pthb,basedir,fname='geneseq',config_file=config_file)\n",
    "    register_seq_images_subsample(pthb,config_pth,chshift_filename,chprofile_filename,cyclename,is_affine,num_initial_c, num_later_c,num_c,subsample_rate,do_coarse,resize_factor, block_size,radius,local,strength,m)\n",
    "    \n",
    "    tn=time.time()\n",
    "    \n",
    "    print(f\"Total time {tn-t0} seconds\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "372ff1d5-6616-44e5-8aef-393e8b8f5bf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_hyb_cycle(pthb,config_pth,basedir,chshift_hyb_filename,chprofile_hyb_filename,config_file='cfg.yaml',cycle_name='n2vhyb',num_c=5,is_affine=0,cyclename_rgb='n2vhyb01',strength=0.6):\n",
    "    t0=time.time()\n",
    "    organize_files(pthb,cyclename='hyb')\n",
    "    denoise_n2v_call(pthb,basedir,fname='hyb',config_file=config_file)\n",
    "    folders,_,_,_=get_folders(pthb)\n",
    "    for i in range(len(folders)):\n",
    "        pthl=os.path.join(pthb,'processed',folders[i])\n",
    "        [shutil.move(f1,os.path.join(pth,'original/'+f1.split('/')[-1])) for f1 in glob.glob(os.path.join(pth,cycle_name+'*.tif'))]\n",
    "        process_hyb_images(pthl,config_pth,chshift_hyb_filename,chprofile_hyb_filename,cycle_name='n2vhyb',ref_cycle_name='alignedn2vgene',reg_cycle=1,dapi_shift=[],dapi_ch_num=5,num_c_hyb=6,num_c=4,is_affine=0,radius=31,ch_i=2)\n",
    "        rgbout(pthb,config_pth,folders[i],cyclename_rgb,shiftmat_name=None,profilemat_name=None,align=0,bleed_correct=0,bcksub=0,m=0,strength=strength,radius=0,num_c=num_c,is_affine=is_affine)\n",
    "    tn=time.time()\n",
    "    print(f\"Total time {tn-t0} seconds\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "0ce718f5-e426-4fc9-820c-8238b5bcb684",
   "metadata": {},
   "outputs": [],
   "source": [
    "def produce_tform_and_stitched_images(pth,sx,sy,intensity_scaling=3,tilesize=3200,transform_rescale_factor=0.5,display_additional_rescale=0.2):\n",
    "    \"\"\"\n",
    "    Stitching function:\n",
    "    1. For all tiles, save the rescaled (downscaled) transformation per tile, apply transformation on RGB of each tile, stitch all the tramsformed RGB into one big image and store it in check_registration folder\n",
    "    2. Calls function to create a merged transform file for global position estimation\n",
    "    \"\"\" \n",
    "    [folders,pos,x,y]=get_folders(pth)\n",
    "    unique_pos=nsort(np.unique(pos))\n",
    "    check_reg_folder=os.path.join(pth,'processed','checkregistration')\n",
    "    os.makedirs(check_reg_folder,exist_ok=True)\n",
    "    folder_names=np.array(folders)\n",
    "    filenames=sorted(glob.glob(os.path.join(pth,'processed',folders[0],'RGB',\"*\"+\".tif\")))\n",
    "    T=load(os.path.join(pth,'tforms_rescaled'+str(transform_rescale_factor).replace('.','p')+'.joblib'))\n",
    "    for f,filename in enumerate(filenames):\n",
    "        filename=filename.split('/')[-1]\n",
    "        I=np.zeros([int(np.max(sy)+tilesize*transform_rescale_factor),int(np.max(sx)+tilesize*transform_rescale_factor),3])\n",
    "        #ax.imshow(I,cmap='jet',vmin=0,vmax=250)\n",
    "        for n_pos in unique_pos:\n",
    "            pos_id=np.array([i for i,name in enumerate(pos) if name==n_pos])\n",
    "            for ids in pos_id:\n",
    "                tilename=folder_names[ids]+'.tif'\n",
    "                tform=skimage.transform.SimilarityTransform(scale=transform_rescale_factor,translation=[T[n_pos][tilename][\"ref_pos\"][0],T[n_pos][tilename][\"ref_pos\"][1]]) \n",
    "                dump(tform,os.path.join(pth,'processed',folder_names[ids],'global_tform_'+str(transform_rescale_factor).replace('.','p')+'.joblib'))\n",
    "                tile=tfl.imread(os.path.join(pth,'processed',folder_names[ids],'RGB',filename))\n",
    "                # if is_grayscale:\n",
    "                #     It=skimage.transform.warp(np.squeeze(tile[stitching_channel-1,:,:]),tform.inverse,preserve_range=True,output_shape=(I.shape[0],I.shape[1]))\n",
    "                # else:\n",
    "                It=skimage.transform.warp(np.squeeze(tile),tform.inverse,preserve_range=True,output_shape=(I.shape[0],I.shape[1]))\n",
    "                I=np.maximum(I,It)\n",
    "   \n",
    "        Irgb=np.uint8(np.clip(I*intensity_scaling,0,255))\n",
    "        Irgb_rescaled = np.uint8(np.clip(rescale(Irgb, display_additional_rescale, channel_axis=-1, preserve_range=True, anti_aliasing=True),0,255))\n",
    "        tfl.imwrite(os.path.join(check_reg_folder,n_pos+'_'+filename),Irgb_rescaled,photometric='rgb')\n",
    "    merge_transforms(pth,str(transform_rescale_factor).replace('.','p'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "00fa2305-4842-405a-81c3-91c2cec32fcf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_transforms(pth,name):\n",
    "    \"\"\"\n",
    "    Stitching function:\n",
    "    Merge all downscaled transforms per tile into a final transformation file\n",
    "    \"\"\" \n",
    "    [folders,_,_,_]=get_folders(pth)\n",
    "    T={}\n",
    "    for folder in folders:\n",
    "        T[folder]=load(os.path.join(pth,'processed',folder,'global_tform_'+name+'.joblib'))\n",
    "    dump(T,os.path.join(pth,'processed','tforms_final.joblib'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "3323b659-d761-4227-93a9-a2a08bf29e6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def stitch_mist_allpos(pth,pth_fiji,fname='n2vgeneseq01',stitching_channel=3,overlap=0.23,numcores=8,is_headless=True,grid_direction='UR',px=0.33,transform_rescale_factor=0.5,num_c=4):\n",
    "    #DON'T KNOW WHY STITCHING CHANNEL IS MENTIONED, MAYBE I WILL COPY ONLY REF CHANNELS LATER, NOT DONE IN THE ORIGINAL MATLAB EITHER\n",
    "    \"\"\"\n",
    "    Stitching function:\n",
    "    1. MIST based stitching--copies all geneseq0 tiles to another stitch folder from where MIST can pick up and stitch\n",
    "    2. For each unique position (slice)-calls MIST for all the tiles belonging to that position (slice)\n",
    "    3. Calls function to merge mist outputs\n",
    "    4. Calls function to rescale the transformations\n",
    "    \"\"\" \n",
    "    [folders,pos,x,y]=get_folders(pth)\n",
    "    unique_pos=nsort(np.unique(pos))\n",
    "    pth_input_stitch=os.path.join(pth,'processed','stitch',fname,'ref_tiles')\n",
    "    pth_output_stitch=os.path.join(pth,'processed','stitch',fname,'stitched')\n",
    "    os.makedirs(pth_input_stitch,exist_ok=True)\n",
    "    os.makedirs(pth_output_stitch,exist_ok=True)\n",
    "    for folder in folders:\n",
    "        shutil.copy(os.path.join(pth,'processed',folder,'original',fname+'.tif'),os.path.join(pth,'processed','stitch',fname,'ref_tiles',folder+'.tif'))\n",
    "    x=np.array(x)\n",
    "    y=np.array(y)\n",
    "    for n_pos in unique_pos:\n",
    "        pos_id=np.array([i for i,name in enumerate(pos) if name==n_pos])\n",
    "        print(pos_id)\n",
    "        grid_x=np.max(x[pos_id])\n",
    "        grid_y=np.max(y[pos_id])\n",
    "        name=folders[pos_id[0]].split('_')[0:2]\n",
    "        outprefix=name[0]+'_'+name[1]+'_'\n",
    "        filename=outprefix+'{ccc}_{rrr}.tif'\n",
    "        stitch_mist_single_pos(grid_x,grid_y,pth_input_stitch,pth_output_stitch,filename,grid_direction,outprefix,px,numcores,pth_fiji,overlap,is_headless)\n",
    "    merge_mist_results(unique_pos,pth_output_stitch,pth)\n",
    "    sx,sy=rescale_transformation(pth,folders,unique_pos,pos,transform_rescale_factor,num_c)\n",
    "    return sx,sy\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "d6f64b64-f8a7-47a5-820a-9a19173d6d9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def stitch_mist_single_pos(grid_x,grid_y,pth_input_stitch,pth_output_stitch,filename,grid_direction,outprefix,px,numcores,pth_fiji,overlap=0.23,is_headless=True):\n",
    "    \"\"\"\n",
    "    Stitching function:\n",
    "    MIST based stitching--per position--calls IJ MIST plugin\n",
    "    \"\"\" \n",
    "    pth_fftwlib=os.path.join(pth_fiji,'lib','fftw')\n",
    "    os.makedirs(pth_fftwlib,exist_ok=True)\n",
    "    pth_fftplan=os.path.join(pth_fftwlib,'fftPlans')\n",
    "    os.makedirs(pth_fftplan,exist_ok=True)\n",
    "    config={\n",
    "        'gridwidth':grid_x,\n",
    "        'gridheight':grid_y,\n",
    "        'starttilerow':0,\n",
    "        'starttilecol':0,\n",
    "        'imagedir':pth_input_stitch,\n",
    "        'filenamepattern':filename,\n",
    "        'filenamepatterntype':'ROWCOL',\n",
    "        'gridorigin':grid_direction,\n",
    "        'assemblefrommetadata':False,\n",
    "        'assemblenooverlap':False,\n",
    "        'globalpositionsfile':[],\n",
    "        'numberingpattern':'HORIZONTALCOMBING',\n",
    "        'startrow':0,\n",
    "        'startcol':0,\n",
    "        'extentwidth':grid_x,\n",
    "        'extentheight':grid_y,\n",
    "        'timeslices':0,\n",
    "        'istimeslicesenabled':False,\n",
    "        'outputpath':pth_output_stitch,\n",
    "        'displaystitching':False,\n",
    "        'outputfullimage':False,\n",
    "        'outputmeta':True,\n",
    "        'outputimgpyramid':False,\n",
    "        'blendingmode':'AVERAGE',\n",
    "        'blendingalpha':0.5,\n",
    "        'compressionmode':'UNCOMPRESSED',\n",
    "        'outfileprefix':outprefix,\n",
    "        'unit':'MICROMETER',\n",
    "        'unitx':px,\n",
    "        'unity':px,\n",
    "        'programtype':'AUTO',\n",
    "        'numcputhreads':numcores,\n",
    "        'loadfftwplan':True,\n",
    "        'savefftwplan':True,\n",
    "        'fftwplantype':'MEASURE',\n",
    "        'ffftwlibraryname':'libfftw3',\n",
    "        'ffftwlibraryfilename':'libfftw3.so',\n",
    "        'planpath':pth_fftplan,\n",
    "        'ffftwlibrarypath':pth_fftwlib,\n",
    "        'stagerepeatability':0,\n",
    "        'horizontaloverlap':overlap*100,\n",
    "        'verticaloverlap':overlap*100,\n",
    "        'numfftpeaks':10,\n",
    "        'overlapuncertainty':'NaN',\n",
    "        'isusedoubleprecision':True,\n",
    "        'isusebioformats':False,\n",
    "        'issuppressmodelwarningdialog':False,\n",
    "        'isenablecudaexceptions':False,\n",
    "        'translationrefinementmethod':'EXHAUSTIVE',\n",
    "        'numtranslationrefinementstartpoints':32,\n",
    "        'headless':is_headless,\n",
    "        'loglevel':'VERBOSE',\n",
    "        'debuglevel':'None'    \n",
    "    }\n",
    "    ij.py.run_plugin(\"MIST\", config)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "45f1446c-33f0-4d1f-acd8-52349e227ab6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_mist_results(unique_pos,pth_output_stitch,pth):\n",
    "    \"\"\"\n",
    "    Stitching function:\n",
    "    1. MIST based stitching results are encoded in a global dictionary \n",
    "    2. For each position (slice)-position of tiles and their names is saved as a sub-dictionary\n",
    "    3. One final dictionary with positions as keys is stored as tforms_original file\n",
    "    \n",
    "    \"\"\" \n",
    "    Texp={}\n",
    "    for n_pos in unique_pos:\n",
    "        lines=[]\n",
    "\n",
    "        Tfull={}\n",
    "        with open(os.path.join(pth_output_stitch,'MAX_'+n_pos+'_global-positions-0.txt'), 'r') as file:\n",
    "            while True:\n",
    "                line = file.readline()\n",
    "                if not line:\n",
    "                    break\n",
    "                lines.append(line.strip())\n",
    "\n",
    "        for line in lines:\n",
    "            T={}\n",
    "            fields=line.split(';')\n",
    "            for field in fields:\n",
    "       \n",
    "                dict_pair=(field.split(':',1))\n",
    "                key=dict_pair[0].strip()\n",
    "                value=dict_pair[-1].strip()\n",
    "                if key=='file':\n",
    "                    tilename=value\n",
    "                elif key=='position':\n",
    "                    T[key]=[int(pos) for pos in value.strip('()').split(',')]\n",
    "                elif key=='grid':\n",
    "                    T[key]=[int(grid) for grid in value.strip('()').split(',')]\n",
    "        \n",
    "            Tfull[tilename]=T\n",
    "            T={}\n",
    "    \n",
    "        Texp[n_pos]=Tfull\n",
    "\n",
    "        Tfull={}\n",
    "    dump(Texp,os.path.join(pth,'processed','tforms_original.joblib'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "11eabae6-ade4-4d88-a7b6-8aa6cd51f3b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_ashlar_results(pth,transform_rescale_factor=0.5,num_c=4):\n",
    "    \"\"\"\n",
    "    Stitching function:\n",
    "    1. ASHLAR based stitching results are encoded in a global dictionary \n",
    "    2. For each position (slice)-position of tiles and their names is saved as a sub-dictionary\n",
    "    3. One final dictionary with positions as keys is stored as tforms_original file\n",
    "    4. Calls function to rescale transformation\n",
    "    \"\"\" \n",
    "    [folders,pos,_,_]=get_folders(pth)\n",
    "    unique_pos=nsort(np.unique(pos))\n",
    "    folder_names=np.array(folders)\n",
    "    Texp={}\n",
    "    Tfull={}\n",
    "    for n_pos in unique_pos:\n",
    "        T={}\n",
    "        df=pd.read_csv(os.path.join(pth,'MAX_'+n_pos+'.positions.tsv'), sep='\\t')\n",
    "        pos_id=np.array([i for i,name in enumerate(pos) if name==n_pos])\n",
    "        for ids in pos_id:\n",
    "            tilename=folder_names[ids]+'.tif'\n",
    "            T['position']=[df.iloc[ids,2],df.iloc[ids,1]]\n",
    "            T['grid']=[0,0]\n",
    "            Tfull[tilename]=T\n",
    "            T={}\n",
    "        Texp[n_pos]=Tfull\n",
    "        Tfull={}\n",
    "    dump(Texp,os.path.join(pth,'processed','tforms_original.joblib'))\n",
    "    sx,sy=rescale_transformation(pth,folders,unique_pos,pos,transform_rescale_factor,num_c)\n",
    "    return sx,sy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "7779950d-ba0a-4c06-8939-066587924a12",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rescale_transformation(pth,folders,unique_pos,pos,rescale_factor=0.5,num_c=4):\n",
    "    \"\"\"\n",
    "    Stitching function:\n",
    "    1. Reads the original transformation dictionary\n",
    "    2. Downscales the coordinates as per rescale_factor and write in as new key-value pairs per tile in the original dictionary\n",
    "    3. Writes the modified dictionary\n",
    "    \n",
    "    \"\"\" \n",
    "\n",
    "    folder_names=np.array(folders)\n",
    "    T=load(os.path.join(pth,'tforms_original.joblib'))\n",
    "    sx=[]\n",
    "    sy=[]\n",
    "    for n_pos in unique_pos:\n",
    "        pos_id=np.array([i for i,name in enumerate(pos) if name==n_pos])\n",
    "        for ids in pos_id:\n",
    "            tilename=folder_names[ids]+'.tif'\n",
    "            T[n_pos][tilename][\"ref_pos\"]=[T[n_pos][tilename][\"position\"][0]*rescale_factor,T[n_pos][tilename][\"position\"][1]*rescale_factor]\n",
    "            sx.append(T[n_pos][tilename][\"ref_pos\"][0])\n",
    "            sy.append(T[n_pos][tilename][\"ref_pos\"][1])\n",
    "        \n",
    "    #pprint.pprint(T)\n",
    "    dump(T,os.path.join(pth,'processed','tforms_rescaled'+str(rescale_factor).replace('.','p')+'.joblib'))\n",
    "    return sx,sy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "b4eda3b8-5e26-4302-8df2-60723784615b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def basecall_hyb_call(pth,thresh,prominence,config_pth,relaxed=1,no_deconv=1,filter_overlap=0,codebook_hyb_name='codebookhyb.mat',cyclename='alignedn2vhyb',num_c=4,all_genes_ch=2):\n",
    "    \"\"\"\n",
    "    Basecalling function:\n",
    "    Calls basecalling hyb function--LOOKS WEIRD TO ME :) WHY WOULD I CODE CALL TO A CALL--I AM TIRED :) :)\n",
    "    \"\"\" \n",
    "    basecall_hyb_all_tiles(pth,cyclename,config_pth,codebook_hyb_name,thresh,prominence,num_c,all_genes_ch,no_deconv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "7dad9f38-c9e0-499e-804d-2dfbe35a368d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def basecall_hyb_all_tiles(pth,cyclename,config_pth,codebook_hyb_name,thresh,prominence,num_c=4,all_genes_ch=2,no_deconv=1):\n",
    "    \"\"\"\n",
    "    Basecalling function:\n",
    "    1. Basecalls hyb for all tiles\n",
    "    2. Writes combined basecall results for all tiles\n",
    "    \"\"\" \n",
    "    # TASK FOR LATER-----THIS PRODUCES NESTED LIST---CHANGE IT BY APPENDING ONLY [0] OF EACH NESTED LIST LIKE BC BASECALL\n",
    "\n",
    "    [folders,_,_,_]=get_folders(pth)\n",
    "    codebook_hyb_path=os.path.join(config_pth,codebook_hyb_name)\n",
    "    codebook_hyb=scipy.io.loadmat(codebook_hyb_path)['codebookhyb']\n",
    "    lroi_x_all=[]\n",
    "    lroi_y_all=[]\n",
    "    id_t_all=[]\n",
    "    sig_t_all=[]\n",
    "    for folder in folders:\n",
    "        hybseq=sorted(glob.glob(os.path.join(pth,'processed',folder,'aligned',cyclename+\"*\"+\"tif\")))\n",
    "        m=len(hybseq)\n",
    "\n",
    "        if m==1:\n",
    "            hyb_raw=tfl.imread(os.path.join(hybseq[0]),key=range(0,num_c,1))\n",
    "            if no_deconv:\n",
    "                hyb_2=hyb_raw\n",
    "                hyb_2[all_genes_ch,:,:]=0\n",
    "                [lroi_x_ind,lroi_y_ind,id_t_ind,sig_t_ind]=basecall_hyb_one_image(os.path.join(pth,'processed',folder,'aligned'),num_c,all_genes_ch,hyb_2,thresh,prominence)\n",
    "            else:\n",
    "                print('Deconvolution to be introduced')\n",
    "        else:\n",
    "            print('Multiple hybseq basecall to be developed')\n",
    "        lroi_x_all.append(lroi_x_ind)\n",
    "        lroi_y_all.append(lroi_y_ind)\n",
    "        id_t_all.append(id_t_ind)\n",
    "        sig_t_all.append(sig_t_ind)\n",
    "        print(f'HYB BASECALL COMPLETE FOR FOLDER {hybseq[0].split(\"/\")[-3]}')\n",
    "    dump({\"lroi_x\":lroi_x_all,\"lroi_y\":lroi_y_all,\"gene_id\":id_t_all,\"signal\":sig_t_all},os.path.join(pth,'processed','genehyb'+'.joblib'))\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "53f3726a-975d-4859-958d-9eec8c7d1cb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "from skimage.morphology import binary_dilation\n",
    "from skimage.measure import regionprops\n",
    "def basecall_hyb_one_image(pthw,num_c,all_genes_ch,hyb_2,thresh,prominence):\n",
    "    \"\"\"\n",
    "    Basecalling function: I FEEL I SHOULD COMBINE MASK AND CALL ONCE-TO REMOVE OVERLAP\n",
    "    1. Basecalls hyb for one tile\n",
    "    2. Writes segmented rolony image per tile\n",
    "    3. Returns basecall results to the calling function\n",
    "    \"\"\" \n",
    "    lroi_x=[]\n",
    "    lroi_y=[]\n",
    "    id_t=[]\n",
    "    sig_t=[]\n",
    "    mask=np.zeros_like(hyb_2)\n",
    "    for n in range(num_c):\n",
    "        if n==all_genes_ch:\n",
    "            mask[n,:,:]=0\n",
    "            continue\n",
    "        else:\n",
    "            a=hyb_2[n,:,:].copy()\n",
    "            a_mask=a>thresh[n]\n",
    "            a_masked=a*a_mask\n",
    "            a_max= extrema.h_maxima(a_masked,prominence[n])\n",
    "            label_peaks = label(a_max)\n",
    "            m=regionprops(label_peaks,a_masked)\n",
    "            mask[n,:,:]=uint16m(binary_dilation(a_max))\n",
    "            [lroi_x,lroi_y,id_t,sig_t]=quantify_peaks(lroi_x,lroi_y,id_t,sig_t,m,hyb_2)\n",
    "\n",
    "    tfl.imwrite(os.path.join(pthw,'mask_hyb.tif'),mask,photometric='minisblack')\n",
    "    return(lroi_x,lroi_y,id_t,sig_t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "1313d285-8e58-45b7-856a-be93eeae7520",
   "metadata": {},
   "outputs": [],
   "source": [
    "def quantify_peaks(lroi_x,lroi_y,id_t,sig_t,m,hyb_2):\n",
    "    \"\"\"\n",
    "    Basecalling function:\n",
    "    1. Based on the regionprops results per tile, this function creates hyb basecalling output and decodes the gene\n",
    "    2. Returns the basecall output to the calling function\n",
    "    \"\"\" \n",
    "    sig1=[]\n",
    "    lroi1_x=[]\n",
    "    lroi1_y=[]\n",
    "    id1=[]\n",
    "    for i,peaks in enumerate(m):\n",
    "        lroi1_x.append(peaks.centroid[0])\n",
    "        lroi1_y.append(peaks.centroid[1])\n",
    "        sig1.append(peaks.intensity_max)\n",
    "        id1.append(np.argmax(hyb_2[:,peaks.coords[0][0],peaks.coords[0][1]])+1) # added 1 here to max channel to match codebook 1,2 and 4\n",
    "    lroi_x.append(lroi1_x)\n",
    "    lroi_y.append(lroi1_y)\n",
    "    id_t.append(id1)\n",
    "    sig_t.append(sig1)\n",
    "    return(lroi_x,lroi_y,id_t,sig_t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "90e2b099-c5d9-4cbd-a040-918ccc02ac7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def basecall_barcodes_rolony(pth,relaxed=0,thresh=[30,30,30,30],prominence=[30,30,30,30],num_cycles=15,num_c=4):\n",
    "    [folders,pos,x,y]=get_folders(pth)\n",
    "    lroi_x_all=[]\n",
    "    lroi_y_all=[]\n",
    "    id_t_all=[]\n",
    "    sig_t_all=[]\n",
    "    score_t_all=[]\n",
    "    seq_t_all=[]\n",
    "    for folder in folders:\n",
    "        print(f'BARCODE BASECALL FOLDER {folder}')\n",
    "        pthw=os.path.join(pth,'processed',folder,'aligned')\n",
    "        I=[]\n",
    "        for i in range(num_cycles):\n",
    "            I.append(tfl.imread(os.path.join(pthw,'alignedregn2vbcseq'+str(\"%0.2d\"%(i+1))+'.tif'),key=range(0,4,1)))\n",
    "        [lroi_x,lroi_y,id_t,sig_t,score_t,seq_t]=basecall_bc_one_image(pthw,num_c,I,thresh,prominence)\n",
    "        lroi_x_all.append(lroi_x[0])\n",
    "        lroi_y_all.append(lroi_y[0])\n",
    "        id_t_all.append(id_t[0])\n",
    "        sig_t_all.append(sig_t[0])\n",
    "        score_t_all.append(score_t[0])\n",
    "        seq_t_all.append(seq_t[0])\n",
    "    score_t_all=np.array(score_t_all)\n",
    "    score_t_all[np.isnan(score_t_all)]=0.5\n",
    "    score_t_all=score_t_all.tolist()\n",
    "    dump({\"lroi_x_all\":lroi_x_all,\"lroi_y_all\":lroi_y_all,\"id_t_all\":id_t_all,\"sig_t_all\":sig_t_all,\"score_t_all\":score_t_all,\"seq_t_all\":seq_t_all},os.path.join(pth,'processed','bc.joblib'))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "8870d25b-61a6-45f3-8345-21d14cd191dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def basecall_bc_one_image(pthw,num_c,I,thresh,prominence):\n",
    "    \"\"\"\n",
    "    Basecalling function:\n",
    "    1. Basecalls bc for one tile\n",
    "    2. Writes segmented rolony image per tile\n",
    "    3. Returns basecall results to the calling function\n",
    "    \"\"\" \n",
    "    lroi_x=[]\n",
    "    lroi_y=[]\n",
    "    id_t=[]\n",
    "    sig_t=[]\n",
    "    score_t=[]\n",
    "    mask=np.zeros_like(I[0])\n",
    "    for n in range(num_c):\n",
    "        a=I[0][n,:,:]\n",
    "        a_mask=a>thresh[n]\n",
    "        a_masked=a*a_mask\n",
    "        a_max= extrema.h_maxima(a_masked,prominence[n])\n",
    "        \n",
    "        \n",
    "        mask[n,:,:]=uint16m(binary_dilation(a_max))# OK, SO I WILL REMOVE DILATION FROM HERE AND DILATE AFTER OVERLAP REMOVAL IN NEXT LINE\n",
    "    comb_mask=clear_overlapping_rolonies(np.max(mask,axis=0))\n",
    "    tfl.imwrite(os.path.join(pthw,'mask_bc.tif'),mask,photometric='minisblack')\n",
    "    tfl.imwrite(os.path.join(pthw,'comb_mask_bc.tif'),comb_mask,photometric='minisblack')\n",
    "    label_peaks = label(comb_mask)\n",
    "    m=regionprops(label_peaks,comb_mask)\n",
    "    \n",
    "    [lroi_x,lroi_y,id_t,sig_t,score_t,seq]=quantify_peaks_bc(lroi_x,lroi_y,id_t,sig_t,score_t,m,I)   \n",
    "\n",
    "    dump({\"lroi_x\":lroi_x,\"lroi_y\":lroi_y,\"id_t\":id_t,\"sig_t\":sig_t,\"score_t\":score_t,\"seq\":seq},os.path.join(pthw,'basecalls-bc.joblib'))\n",
    "\n",
    "    return(lroi_x,lroi_y,id_t,sig_t,score_t,seq)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "e9843145-1bc2-4545-a99d-c022f7f4d4df",
   "metadata": {},
   "outputs": [],
   "source": [
    "def clear_overlapping_rolonies(a):\n",
    "    c=np.array([[0,1,1],[0,0,1],[0,0,1]])\n",
    "    a=a & skimage.util.invert(binary_dilation(a,c))\n",
    "    return a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "a866a167-12a3-4add-9ed5-07e472986a71",
   "metadata": {},
   "outputs": [],
   "source": [
    "def quantify_peaks_bc(lroi_x,lroi_y,id_t,sig_t,score_t,m,I):\n",
    "    \"\"\"\n",
    "    Basecalling function:\n",
    "    1. Based on the regionprops results per tile, this function creates bc basecalling output and decodes the gene\n",
    "    2. Returns the basecall output to the calling function\n",
    "    \"\"\" \n",
    "    sig2=[]\n",
    "    lroi1_x=[]\n",
    "    lroi1_y=[]\n",
    "    id2=[]\n",
    "    score2=[]\n",
    "    gene_map=np.array(list(\"GTAC\")) \n",
    "    \n",
    "    for i,peaks in enumerate(m):\n",
    "        lroi1_x.append(peaks.centroid[0])\n",
    "        lroi1_y.append(peaks.centroid[1])\n",
    "        score1=[]\n",
    "        sig1=[]\n",
    "        id1=[]\n",
    "        #score=0\n",
    "        for j in range(len(I)):\n",
    "            intensity=I[j][:,peaks.coords[0][0],peaks.coords[0][1]]\n",
    "            #score[np.isnan(score)]=0.5\n",
    "            sig1.append(np.max(intensity))\n",
    "            id1.append(np.argmax(intensity))\n",
    "            score1.append(np.max(intensity)/np.sqrt(np.sum(np.square(intensity))))\n",
    "        sig2.append(sig1)\n",
    "        id2.append(id1)\n",
    "        score2.append(score1)\n",
    "    lroi_x.append(lroi1_x)\n",
    "    lroi_y.append(lroi1_y)\n",
    "    id_t.append(id2)\n",
    "    sig_t.append(sig2)\n",
    "    score_t.append(score2)\n",
    "    seq=gene_map[id_t]   \n",
    "    return(lroi_x,lroi_y,id_t,sig_t,score_t,seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "d33805e6-c449-467d-ab46-e6d7418b8b7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def basecall_bc_soma_all(pth,num_ch=4,mname='dil_cell_mask_cyto3.tif',fname='alignedregn2vbcseq'):\n",
    "    bc_label_all=[]\n",
    "    bc_sig_all_channels_all=[]\n",
    "    bc_sig_all=[]\n",
    "    bc_id_all=[]\n",
    "    bc_score_all=[]\n",
    "    bc_seq_all=[]\n",
    "    [folders,pos,x,y]=get_folders(pth)\n",
    "\n",
    "    for folder in folders:\n",
    "        [bc_label,bc_sig_all_channels,bc_sig,bc_id,bc_score,bc_seq]=basecall_bc_soma_one_image(pth,folder,num_ch,mname,fname)\n",
    "        bc_sig_all_channels_all.append(bc_sig_all_channels)\n",
    "        bc_sig_all.append(bc_sig)\n",
    "        bc_id_all.append(bc_id)\n",
    "        bc_score_all.append(bc_score)\n",
    "        bc_seq_all.append(bc_seq)\n",
    "        bc_label_all.append(bc_label)\n",
    "        print(f'Basecalled soma-bc folder {folder}')\n",
    "\n",
    "    dump({\"bc_label\":bc_label_all,\"bc_sig_all_channels\":bc_sig_all_channels_all,\"bc_sig\":bc_sig_all,\" bc_id\": bc_id_all,\"bc_score\":bc_score_all,\"bc_seq\":bc_seq_all},os.path.join(pth,'processed','all_bccells_intensity.joblib'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "edd10144-eaf7-49df-983c-54bd0c7d6489",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "def basecall_bc_soma_one_image(pth,folder,num_ch=4,mname='dil_cell_mask_cyto3.tif', fname='alignedregn2vbcseq'):\n",
    "    # NUCLEAR PROFILE BASED BACKGROUND SUBTRACTION NOT DONE\n",
    "    gene_map=np.array(list(\"GTAC\"))   \n",
    "    mask=tfl.imread(os.path.join(pth,'processed',folder,'aligned',mname))\n",
    "    bc_label_cycles=[]\n",
    "    bc_sig_all_channels_cycles=[]\n",
    "    bc_sig_cycles=[]\n",
    "    bc_id_cycles=[]\n",
    "    bc_score_cycles=[]\n",
    "    bc_seq_cycles=[]\n",
    "    if np.max(mask[:])>0:\n",
    "        for i in range(num_cycles):\n",
    "            bc_label=[]\n",
    "            bc_sig_all_channels=[]\n",
    "            bc_sig=[]\n",
    "            bc_id=[]\n",
    "            bc_score=[]\n",
    "            bc_seq=[]\n",
    "            I=np.transpose(tfl.imread(os.path.join(pth,'processed',folder,'aligned',fname+str(\"%0.2d\"%(i+1))+'.tif'),key=range(0,num_ch,1)), axes=(1,2,0))\n",
    "            cell_data=regionprops_table(mask,I,properties=('label','intensity_mean'))\n",
    "            bc_label=cell_data['label']\n",
    "            bc_sig_all_channels=[cell_data['intensity_mean-0'],cell_data['intensity_mean-1'],cell_data['intensity_mean-2'],cell_data['intensity_mean-3']] # [channel][cells]\n",
    "            bc_sig=np.max(np.array(bc_sig_all_channels),axis=0)\n",
    "            bc_id=np.argmax(np.array(bc_sig_all_channels),axis=0)\n",
    "            bc_score=bc_sig/np.sqrt(np.sum(np.square(bc_sig_all_channels),axis=0))\n",
    "            bc_score[np.isnan(bc_score)]=0.5\n",
    "            bc_seq=gene_map[bc_id]\n",
    "        \n",
    "            bc_label_cycles=bc_label\n",
    "            bc_sig_all_channels_cycles.append(bc_sig_all_channels)\n",
    "            bc_sig_cycles.append(bc_sig)\n",
    "            bc_id_cycles.append(bc_id)\n",
    "            bc_score_cycles.append(bc_score)\n",
    "            bc_seq_cycles.append(bc_seq)\n",
    "            \n",
    "    dump({\"bc_label\":bc_label_cycles,\"bc_sig_all_channels\":bc_sig_all_channels_cycles,\"bc_sig\":bc_sig_cycles,\" bc_id\": bc_id_cycles,\"bc_score\":bc_score_cycles,\"bc_seq\":bc_seq_cycles},os.path.join(pth,'processed',folder,'aligned','bc-somas.joblib'))\n",
    "    \n",
    "    return bc_label_cycles,bc_sig_all_channels_cycles,bc_sig_cycles,bc_id_cycles,bc_score_cycles,bc_seq_cycles\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "d1bf6af0-da55-44ee-af05-e43cdf2c3d7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_codebook_bin(pth,num_channels=4,codebook_name='codebookM1all.mat'):\n",
    "    \"\"\"\n",
    "    Basecalling function for geneseq:\n",
    "    1. In preparation for bardensr-this function creates binarized codebook as per the number of geneseq cycles in the experiment\n",
    "    \"\"\" \n",
    "    [folders,_,_,_]=get_folders(pth)\n",
    "    genecycles=len(sorted(glob.glob(os.path.join(pth,'processed',folders[0],'original',\"*geneseq*\"))))    \n",
    "    codebook=scipy.io.loadmat(os.path.join(config_pth,codebook_name))['codebook']\n",
    "    genes=np.reshape(np.array([str(x[0][0]) for x in codebook]),(np.size(codebook,0),-1))\n",
    "    codebook1=np.zeros((np.size(codebook,0),genecycles),dtype=str)\n",
    "\n",
    "    for i in range(np.size(codebook,0)):\n",
    "        for j in range(genecycles):\n",
    "            codebook1[i,j]=codebook[i][1][0][j].copy() # CHECK IF COPY IS OK\n",
    "\n",
    "    codebook_bin=np.ones(np.shape(codebook1),dtype=np.double)\n",
    "    codebook_bin=np.reshape(np.array([float(x.replace('G','8').replace('T','4').replace('A','2').replace('C','1')) for y in codebook1 for x in y]),np.shape(codebook1))\n",
    "\n",
    "    codebook_bin=np.matmul(np.uint8(codebook_bin),2**np.transpose(np.array((np.arange(4*genecycles-4,-1,-4)))))\n",
    "\n",
    "    codebook_bin=np.array([bin(i)[2:].zfill(genecycles*num_channels) for i in codebook_bin])\n",
    "    codebook_bin=np.reshape([np.uint8(i) for j in codebook_bin for i in j],(np.size(codebook1,0),genecycles*num_channels))\n",
    "\n",
    "    co=[[genes[i],codebook_bin[j,:]] for i in range(np.size(genes,0))]\n",
    "    co=[codebook,co]\n",
    "    codebook_bin1=np.reshape(codebook_bin,(np.size(codebook_bin,0),-1,num_channels))\n",
    "    dump(co,os.path.join(pth,'processed','codebook.joblib'))\n",
    "    dump(codebook_bin1,os.path.join(pth,'processed','codebookforbardensr.joblib'))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "c0de9f34-c336-4006-a0a7-c6ae35e412c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def bardensr_call(pth,config_pth,num_channels=4,codebook_name='codebookM1all.mat',fdrthresh=0.05,trim=160,cropf=0.4,noisefloor_ini=0.01,noisefloor_final=0.05):\n",
    "    \"\"\"\n",
    "    Basecalling function for geneseq:\n",
    "    1. Calls the function that creates binarized codebook as per the number of geneseq cycles in the experiment\n",
    "    2. Calculates normalization factors, intensity thresholds for spot detection based on allowed fdr\n",
    "    3. Calls Bardensr spot calling and peak detection to identify and decode rolonies per tile\n",
    "    4. Saves basecalling results per tile\n",
    "    \"\"\" \n",
    "    make_codebook_bin(pth,num_channels,codebook_name)\n",
    "\n",
    "    controlidx=range(25,35)\n",
    "    [folders,_,_,_]=get_folders(pth)\n",
    "    codebook=scipy.io.loadmat(os.path.join(config_pth,codebook_name))['codebook']\n",
    "    genes=np.array([str(x[0][0]) for x in codebook],dtype=str)\n",
    "    cb=load(os.path.join(pth,'processed','codebookforbardensr.joblib'))\n",
    "    cb=np.transpose(cb,axes=(1,2,0))\n",
    "    R,C,J=cb.shape\n",
    "    codeflat=np.reshape(cb,(-1,J))\n",
    "    pos_unused_codes=np.where(np.char.startswith(genes,'unused'))\n",
    "    err_codes=genes[pos_unused_codes]\n",
    "\n",
    "    # SELECTING 10 FOLDERS AS CONTROL\n",
    "    if len(folders)>=36:\n",
    "        control_folders=folders[controlidx]\n",
    "    else:\n",
    "        control_folders=folders\n",
    "\n",
    "    # NORMALIZATION-PREPROCESSING--CALCULATING MAX OF EACH CYCLE AND EACH CHANNEL ACROSS ALL CONTROL FOVS\n",
    "    max_per_RC=[image_reader_cropped(os.path.join(pth,'processed',i,'aligned'),R,C,cropf).max(axis=(1,2,3)) for i in control_folders]\n",
    "    median_max=np.median(max_per_RC,axis=0)\n",
    "\n",
    "    # ESTABLISHING BASE THRESHOLD AT THE MEDIAN OF MAXIMUM ERROR READOUT \n",
    "\n",
    "    err_max=[]\n",
    "    evidence_tensors=[]\n",
    "    for folder in control_folders:\n",
    "        et=bardensr.spot_calling.estimate_density_singleshot(image_reader_trimmed(os.path.join(pth,'processed',folder,'aligned'),trim,R,C)/median_max[:,None,None,None],codeflat,noisefloor_ini)\n",
    "        err_max.append(et[:,:,:,pos_unused_codes].max(axis=(0,1,2)))\n",
    "    err_max=np.array(err_max)\n",
    "    thresh=np.median(np.median(err_max,axis=1))\n",
    "    print(thresh)\n",
    "\n",
    "    # FINDING OPTIMUM TRHESHOLD WITH LOWEST FDR ON CONTROL FOLDERS\n",
    "\n",
    "    err_c_all=[]\n",
    "    total_c_all=[]\n",
    "    for folder in control_folders:\n",
    "        et=bardensr.spot_calling.estimate_density_singleshot(image_reader_cropped(os.path.join(pth,'processed',folder,'aligned'),R,C,cropf)/median_max[:,None,None,None],codeflat,noisefloor_final)\n",
    "        for thresh1 in np.linspace(thresh-0.1,thresh+0.1,10):\n",
    "            spots=bardensr.spot_calling.find_peaks(et,thresh1,use_tqdm_notebook=False)\n",
    "            spots.to_csv(os.path.join(pth,'processed',folder,'aligned','spots.csv'),index=False)\n",
    "            print(f\"in {folder} found {len(spots)} spots\")\n",
    "            err_c=0\n",
    "            for err_idx in pos_unused_codes[0]:\n",
    "                err_c=err_c+(spots.j==err_idx).to_numpy().sum()\n",
    "            err_c_all.append(err_c)\n",
    "            total_c_all.append(len(spots)-err_c)      \n",
    "    #calculate fdr        \n",
    "    err_c_all1=np.reshape(err_c_all,[len(control_folders),10])\n",
    "    total_c_all1=np.reshape(total_c_all,[len(control_folders),10])+1\n",
    "    fdr=err_c_all1/len(pos_unused_codes[0])*(len(genes)-len(pos_unused_codes[0]))/(total_c_all1)\n",
    "    fdrmean=err_c_all1.mean(axis=0)/len(pos_unused_codes[0])*(len(genes)-len(pos_unused_codes[0]))/(total_c_all1.mean(axis=0))\n",
    "    thresh_refined=np.linspace(thresh-0.1,thresh+0.1,10)[(fdrmean<fdrthresh).nonzero()[0][0]]#this is the new threshold optimized by targeted fdr value\n",
    "    print(thresh_refined)  \n",
    "    with open(os.path.join(pth,'processed','thresh_refined.txt'),'w') as f:\n",
    "        f.write(str(thresh_refined))\n",
    "    with open(os.path.join(pth,'processed','noise_floors.txt'),'w') as f:\n",
    "        f.write(str(noisefloor_final))\n",
    "    print(f\"threshold {thresh_refined} with noise floor {noisefloor_final}\")\n",
    "\n",
    "    # MAIN BASE-CALLING ON ALL FOLDERS-TRIMMED IMAGES WITH FINALIZED THRESHOLD\n",
    "    # # spot call each fov, using the thresholds we decided on, and the normalization we decided on\n",
    "    for folder in folders:\n",
    "        et=bardensr.spot_calling.estimate_density_singleshot(image_reader_trimmed(os.path.join(pth,'processed',folder,'aligned'),trim,R,C)/median_max[:,None,None,None],codeflat,noisefloor_final)\n",
    "        spots=bardensr.spot_calling.find_peaks(et,thresh_refined,use_tqdm_notebook=False)\n",
    "        spots.loc[:,'m1']=spots.loc[:,'m1']+trim\n",
    "        spots.loc[:,'m2']=spots.loc[:,'m2']+trim\n",
    "        spots.to_csv(os.path.join(pth,'processed',folder,'aligned','bardensrresult.csv'),index=False)\n",
    "        print(f\"in {folder} found {len(spots)} spots\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "b95b5d5f-bd49-402d-a24b-664786c831fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def import_bardensr_results(pth,fname='bardensrresult.csv',prev_codebook_len=0,codebook_name='codebook.joblib',is_optseq=0,codebook_opt_name='codebook.mat'):\n",
    "    \"\"\"\n",
    "    Basecalling function for geneseq:\n",
    "    1. Calls an accumulator function to aggregate bardensr results per tile\n",
    "    2. Calls function to compute fdr for all tiles combined\n",
    "    3. Writes aggregated geneseq basecall results\n",
    "    \"\"\"\n",
    "    lroi_x=[] \n",
    "    lroi_y=[]\n",
    "    gene_id=[]\n",
    "\n",
    "    [lroi_x,lroi_y,gene_id]=accumulate_bardensr_results(pth,codebook_name,fname,0)\n",
    "    fdr=get_fdr(pth,codebook_name,gene_id)\n",
    "    print(f'Finished importing bardensr results. FPR is {fdr}')\n",
    "\n",
    "    if is_optseq:\n",
    "        print('Has optseq')\n",
    "        codebook_path=os.path.join(pth,'processed',codebook_opt_name)\n",
    "        codebook=scipy.io.loadmat(codebook_path)['codebook']\n",
    "        [lroi_x_opt,lroi_y_opt,gene_id_opt]=accumulate_bardensr_results(pth,codebook_opt_name,fname_opt,len(codebook))\n",
    "        fdr_o=get_fdr(pth,codebook_opt_name,gene_id_opt)\n",
    "        gene_id.append(gene_id_opt)\n",
    "        lroi_x.append(lroi_x_opt)\n",
    "        lroi_y.append(lroi_y_opt)\n",
    "    dump({\"lroi_x\":lroi_y,\"lroi_y\":lroi_x,\"gene_id\":gene_id},os.path.join(pth,'processed','basecalls'+'.joblib')) # let's check if this works --exchanging x and y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "e89458f3-d244-421f-ba3a-505f186153d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def accumulate_bardensr_results(pth,codebook_name,fname,prev_codebook_len):\n",
    "    \"\"\"\n",
    "    Basecalling function for geneseq:\n",
    "    1. Accumulator function that combines bardensr results per tile\n",
    "    2. Returns combined basecall results to the calling function\n",
    "    3. WHY AM I READING CODEBOOK HERE?----------------CHECK AGAIN-----------------\n",
    "    \"\"\"\n",
    "    # codebook_path=os.path.join(pth,'processed',codebook_name)\n",
    "    # codebook=scipy.io.loadmat(codebook_path)['codebook']\n",
    "\n",
    "    [folders,_,_,_]=get_folders(pth)\n",
    "    lroi_x=[] \n",
    "    lroi_y=[]\n",
    "    gene_id=[]\n",
    "\n",
    "    for i,folder in enumerate(folders):\n",
    "        try:\n",
    "            T=pd.read_csv(os.path.join(pth,'processed',folder,'aligned',fname),header=0)\n",
    "            lroi_x.append(T.m2)\n",
    "            lroi_y.append(T.m1)\n",
    "            gene_id.append(T.j+prev_codebook_len)\n",
    "        except:\n",
    "            lroi_x.append([])\n",
    "            lroi_y.append([])\n",
    "            gene_id.append([])\n",
    "            print(f'No geneseq rolonies found for tile {folder}\\n')\n",
    "\n",
    "    return(lroi_x,lroi_y,gene_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "76fec2b7-9880-4c0f-8630-19516a096301",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_fdr(pth,codebook_name,gene_id):\n",
    "    \"\"\"\n",
    "    Basecalling function for geneseq:\n",
    "    Calculates fdr\n",
    "    \"\"\"\n",
    "    codebook_path=os.path.join(pth,'processed',codebook_name)\n",
    "    #codebook=scipy.io.loadmat(codebook_path)['codebook']\n",
    "    codebook_full=load(codebook_path)\n",
    "    codebook=codebook_full[0]\n",
    "    genes=np.array([str(x[0][0]) for x in codebook],dtype=str)\n",
    "    pos_unused_codes=np.where(np.char.startswith(genes,'unused'))\n",
    "    err_codes=genes[pos_unused_codes]\n",
    "    id_flat = np.array([code for pos in gene_id for code in pos])\n",
    "    err_c_all1=id_flat[np.isin(id_flat,pos_unused_codes[0])]\n",
    "    fdr=(len(err_c_all1)/len(id_flat))/len(pos_unused_codes[0])*(len(genes)-len(pos_unused_codes[0]))\n",
    "    return(fdr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "e6bac02d-24c2-4bba-ab66-67504e2cb5e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def image_reader_cropped(pthb,R,C,cropf=0.4):\n",
    "    \"\"\"\n",
    "    Helper function for basecalling function for geneseq:\n",
    "    Returns a nearly center cropped in xy image stack (with cycles and channels intact)\n",
    "    \"\"\"\n",
    "    I=[]\n",
    "    for i in range(1,R+1):\n",
    "        for j in range(C):\n",
    "            I.append(np.expand_dims(tfl.imread(os.path.join(pthb,'alignedn2vgeneseq'+'%0.2d'%i+'.tif'),key=j),axis=0))\n",
    "    I=np.array(I)\n",
    "    nx=np.size(I,3)\n",
    "    ny=np.size(I,2)\n",
    "    return(I[:,:,round(ny*cropf):round(ny*(1-cropf)),round(nx*cropf):round(nx*(1-cropf))])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "fc17e7a0-826a-46ca-b356-d3828ae5861e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def image_reader_trimmed(pthb,trim,R,C):\n",
    "    \"\"\"\n",
    "    Helper function for basecalling function for geneseq:\n",
    "    Returns a border trimmed in xy image stack (with cycles and channels intact)\n",
    "    \"\"\"\n",
    "    I=[]\n",
    "    for i in range(1,R+1):\n",
    "        for j in range(C):\n",
    "            I.append(np.expand_dims(tfl.imread(os.path.join(pthb,'alignedn2vgeneseq'+'%0.2d'%i+'.tif'),key=j),axis=0))\n",
    "    I=np.array(I)\n",
    "\n",
    "    return(I[:,:,trim:-trim,trim:-trim])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "c3c489c4-8a63-4b00-b852-5cded4a33730",
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_cellpose_input(pth,nuc_ch=5,num_chyb=5,num_cgene=4,tilesize=3200):\n",
    "    \"\"\"\n",
    "    Cell segmentation function:\n",
    "    1. creates an input image stack for cellpose input per tile with nuclear  information from DAPI (hyb) and gene from rest of the hyb and all geneseqs\n",
    "    \"\"\"\n",
    "    [folders,pos,x,y]=get_folders(pth)\n",
    "    other_ch=list(range(0,num_chyb))\n",
    "    del other_ch[nuc_ch-1]\n",
    "    cell_inp=np.zeros([2,tilesize,tilesize])\n",
    "    for folder in folders:\n",
    "        Ih=tfl.imread(os.path.join(pth,'processed',folder,'aligned','alignedn2vhyb01.tif'),key=range(0,num_chyb,1))\n",
    "        Ig=np.zeros([Ih.shape[1],Ih.shape[2]])\n",
    "        gene_cycles=sorted(glob.glob(os.path.join(pth,'processed',folder,'aligned','alignedfixedn2vgeneseq*.tif')))\n",
    "        for gene_cycle in gene_cycles:\n",
    "            Ig=Ig+np.sum(tfl.imread(gene_cycle,key=range(0,num_cgene,1)),axis=0)\n",
    "        Inuc=Ih[nuc_ch-1,:,:]\n",
    "        Icyto=np.sum(Ih[other_ch,:,:],axis=0)+Ig\n",
    "        cell_inp[0,:,:]=uint16m(Icyto)\n",
    "        cell_inp[1,:,:]=uint16m(Inuc)\n",
    "        tfl.imwrite(os.path.join(pth,'processed',folder,'aligned','cell_inp2.tif'),uint16m(cell_inp),photometric='minisblack')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "579ca488-8786-48f3-bcae-bfa87ae63faf",
   "metadata": {},
   "outputs": [],
   "source": [
    "from skimage.segmentation import expand_labels\n",
    "from skimage.measure import label,regionprops_table\n",
    "\n",
    "def import_single_tile_segmentation(pth,folder,dilation_radius=3,fname='cell_mask_cyto3.tif'):\n",
    "    \"\"\"\n",
    "    Cell segmentation function:\n",
    "    1. Reads the output from cellpose per tile--the label mask image\n",
    "    2. Dilates it by few pixels to increase cellular coverage\n",
    "    3. Assigns properties to each segmented cell and returns these properties per tile to the calling function\n",
    "    \"\"\"\n",
    "    mask=tfl.imread(os.path.join(pth,'processed',folder,'aligned',fname))\n",
    "    mask_dil=expand_labels(mask,dilation_radius)\n",
    "    measure=regionprops_table(mask_dil,properties=('label','centroid'))\n",
    "    cell_num=measure['label']\n",
    "    cent_x=measure['centroid-0']\n",
    "    cent_y=measure['centroid-1']\n",
    "\n",
    "    return mask,mask_dil,cell_num,cent_x,cent_y\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "ad6e0968-2b9e-4ead-a6c8-afaa02bc9e86",
   "metadata": {},
   "outputs": [],
   "source": [
    "def import_cellpose_all_tiles(pth,dilation_radius=3,fname='cell_mask_cyto3.tif'):\n",
    "    \"\"\"\n",
    "    Cell segmentation function:\n",
    "    1. Calls single tile fucntion to measure properties of cells--aggregates them for all tiles\n",
    "    2. Writes the combined segmentation result for all tiles\n",
    "    \"\"\"\n",
    "    [folders,_,_,_]=get_folders(pth)\n",
    "    all_segmentation_dict={}\n",
    "    for i,folder in enumerate(folders):\n",
    "        tile_data={}\n",
    "        [mask,mask_dil,cell_num,cent_x,cent_y]=import_single_tile_segmentation(pth,folder,dilation_radius,fname)\n",
    "        tile_data['original_labels']=mask\n",
    "        tile_data['dilated_labels']=mask_dil\n",
    "        tile_data['cell_num']=cell_num\n",
    "        tile_data['cent_x']=cent_x\n",
    "        tile_data['cent_y']=cent_y\n",
    "        all_segmentation_dict[folder]=tile_data\n",
    "        io.imsave(os.path.join(pth,'processed',folder,'aligned','dil_'+fname),mask_dil)\n",
    "    dump(all_segmentation_dict,os.path.join(pth,'processed','all_segmentation.joblib'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "4e8f0ba9-338c-4509-8a89-5fe21e742734",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cellid(mask,coord_x,coord_y):\n",
    "    \"\"\"\n",
    "    Global transformation function:\n",
    "    1. For any detected rolony-assigns it to a cell\n",
    "    2. Returns the cell ids for all rolonies in this tile\n",
    "    \"\"\"\n",
    "    coord_xl=[int(np.round(x)) for x in coord_x]\n",
    "    coord_yl=[int(np.round(x)) for x in coord_y]\n",
    "    cell_id=mask[coord_xl,coord_yl]\n",
    "    return cell_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "c93afabf-c3cf-4b38-826a-4b2f95beca1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def assign_rolony_to_cell(mask,coord_x,coord_y):\n",
    "    \"\"\"\n",
    "    Global transformation function:\n",
    "    1. Calls get_cellid function if there are rolonies detected in this tile or else assigns empty cell id to this tile\n",
    "    \"\"\"\n",
    "    if len(coord_x):\n",
    "        cell_id=get_cellid(mask,coord_x,coord_y)\n",
    "    else:\n",
    "        cell_id=[] # earlier this was [] and was causing error later\n",
    "    return cell_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "12f311c8-4d8b-4241-a199-21bfad7b6fa1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def aggregate_cell_ids_gene_hyb(pth,is_bc):\n",
    "    \"\"\"\n",
    "    Global transformation function:\n",
    "    1. Combines gene and hyb basecalls and assigns original per tile position and cell id to each\n",
    "    2. Writes the combined rolony properties file\n",
    "    \"\"\"\n",
    "    gene_rol=load(os.path.join(pth,'processed','basecalls.joblib'))\n",
    "    seg=load(os.path.join(pth,'processed','all_segmentation.joblib'))\n",
    "    hyb_rol=load(os.path.join(pth,'processed','genehyb.joblib'))\n",
    "    if is_bc:\n",
    "        bc_rol=load(os.path.join(pth,'processed','bc.joblib'))\n",
    "    \n",
    "    [folders,_,_,_]=get_folders(pth)\n",
    "    T={}\n",
    "    Tbc={}\n",
    "    for i,folder in enumerate(folders):\n",
    "        print(f'Operating on {folder}')\n",
    "        t={}\n",
    "        mask=seg[folder]['dilated_labels']\n",
    "        coord_xg=gene_rol['lroi_x'][i]\n",
    "        coord_yg=gene_rol['lroi_y'][i]\n",
    "        coord_xh=hyb_rol['lroi_x'][i][0]\n",
    "        coord_yh=hyb_rol['lroi_y'][i][0]\n",
    "        t['cellid']=assign_rolony_to_cell(mask,coord_xg,coord_yg)\n",
    "        t['cellidhyb']=assign_rolony_to_cell(mask,coord_xh,coord_yh)\n",
    "        T[folder]=t\n",
    "        if is_bc:\n",
    "            tbc={}\n",
    "            coord_xb=bc_rol['lroi_x_all'][i][0]\n",
    "            coord_yb=bc_rol['lroi_y_all'][i][0]\n",
    "            tbc['cellidbc']=assign_rolony_to_cell(mask,coord_xb,coord_yb)\n",
    "            Tbc[folder]=tbc\n",
    "    dump(T,os.path.join(pth,'processed','cell_id.joblib'))\n",
    "    if is_bc:\n",
    "        dump(Tbc,os.path.join(pth,'processed','bccellid.joblib'))\n",
    "    print('ALL ROLONIES ASSIGNED TO CELLS')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "89381c52-801b-439a-b5fe-8b94feb64c03",
   "metadata": {},
   "outputs": [],
   "source": [
    "def transform_downsized_coordinates(pth,is_bc):\n",
    "    gene_rol=load(os.path.join(pth,'processed','basecalls.joblib'))\n",
    "    seg=load(os.path.join(pth,'processed','all_segmentation.joblib'))\n",
    "    hyb_rol=load(os.path.join(pth,'processed','genehyb.joblib'))\n",
    "    Tf=load(os.path.join(pth,'processed','tforms_final.joblib'))\n",
    "    if is_bc:\n",
    "        bc_rol=load(os.path.join(pth,'processed','bc.joblib'))\n",
    "    [folders,_,_,_]=get_folders(pth)\n",
    "    T={}\n",
    "    for i,folder in enumerate(folders):\n",
    "        t={}\n",
    "        tform=Tf[folder]\n",
    "        [x,y]=apply_transform(tform,gene_rol['lroi_y'][i],gene_rol['lroi_x'][i])\n",
    "        t['lroi10x_x']=x\n",
    "        t['lroi10x_y']=y\n",
    "        [x,y]=apply_transform(tform,hyb_rol['lroi_y'][i][0],hyb_rol['lroi_x'][i][0]) \n",
    "        t['lroi10xhyb_x']=x\n",
    "        t['lroi10xhyb_y']=y\n",
    "        [x,y]=apply_transform(tform,seg[folder]['cent_y'],seg[folder]['cent_x']) \n",
    "        t['cellpos10x_x']=x\n",
    "        t['cellpos10x_y']=y\n",
    "        if is_bc:\n",
    "            [x,y]=apply_transform(tform,bc_rol['lroi_y_all'][i][0],bc_rol['lroi_x_all'][i][0]) \n",
    "            t['bcpos10x_x']=x\n",
    "            t['bcpos10x_y']=y\n",
    "        T[folder]=t\n",
    "    dump(T,os.path.join(pth,'processed','lroi10x.joblib'))\n",
    "    print('ALL COORDINATES TRANSFORMED')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "d54bec61-27b8-494f-8c8c-b145cdf18eb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def apply_transform(tform,coord_x,coord_y):\n",
    "    \"\"\"\n",
    "    Global transformation function:\n",
    "    1. Transforms the local coordinates of rolonies and cells to global downsized coordinates per tile\n",
    "    \"\"\"\n",
    "    \n",
    "    if len(coord_x):\n",
    "        if not (isinstance(coord_x,list) or isinstance(coord_x,np.ndarray)):\n",
    "            coord_x=coord_x.to_list()\n",
    "            coord_y=coord_y.to_list()\n",
    "        q=np.zeros([len(coord_x),2])\n",
    "        q[:,0]=np.reshape(coord_x,(1,-1))\n",
    "        q[:,1]=np.reshape(coord_y,(1,-1))\n",
    "        v=tform(q)\n",
    "        x=v[:,0]\n",
    "        y=v[:,1]\n",
    "    else:\n",
    "        x=[]\n",
    "        y=[]\n",
    "        \n",
    "    return x,y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "86a9d0b8-5598-412f-9996-8d2287ecd044",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.sparse import coo_matrix\n",
    "import datetime\n",
    "\n",
    "\n",
    "def organize_processed_data(pth,config_pth,is_optseq=0,hyb_codebook_name='codebookhyb.mat',starting_slice_idx=1,starting_fov_idx=1,dummy_cell_num=10000,tilesize=3200,\n",
    "                           fraction_border=0.07):\n",
    "    \"\"\"\n",
    "    Postprocessing function:\n",
    "    Organizes all the data into rolonies and neurons--prepares a big dictionary\n",
    "    \"\"\"\n",
    "    today=datetime.date.today().strftime('%d%m%Y')\n",
    "    gene_rol=load(os.path.join(pth,'processed','basecalls.joblib'))\n",
    "    seg=load(os.path.join(pth,'processed','all_segmentation.joblib'))\n",
    "    hyb_rol=load(os.path.join(pth,'processed','genehyb.joblib'))\n",
    "    Tf=load(os.path.join(pth,'processed','tforms_final.joblib'))\n",
    "    coord=load(os.path.join(pth,'processed','lroi10x.joblib'))\n",
    "    codebook=load(os.path.join(pth,'processed','codebook.joblib'))\n",
    "    cellid=load(os.path.join(pth,'processed','cell_id.joblib'))\n",
    "    hyb_codebook=scipy.io.loadmat(os.path.join(config_pth,hyb_codebook_name))['codebookhyb']\n",
    "    dump([hyb_codebook],os.path.join(pth,'processed','hyb_codebook.joblib'))\n",
    "    hyb_codebook=load(os.path.join(pth,'processed','hyb_codebook.joblib'))\n",
    "    [folders,pos,_,_]=get_folders(pth)\n",
    "    npos=nsort(np.unique(pos))\n",
    "    d={}\n",
    "    d=data_dict_organizer(d,'initialize',fov=[],gene_rol_id=[],pos_10x_allx=[],pos_10x_ally=[],pos_40x_allx=[],pos_40x_ally=[],\n",
    "                          cellidall=[],sliceidall=[],hyb_rol_id=[],fov_hyb=[],pos_10x_allx_hyb=[],pos_10x_ally_hyb=[],pos_40x_allx_hyb=[],pos_40x_ally_hyb=[],\n",
    "                          cellidall_hyb=[],sliceidall_hyb=[],cell_list_all=[],cell_pos_10x_allx=[],cell_pos_10x_ally=[],cell_pos_40x_allx=[],cell_pos_40x_ally=[],\n",
    "                          fov_cell=[],sliceidall_cell=[])\n",
    "\n",
    "    for i,folder in enumerate(folders):\n",
    "        pos_id=np.array([j for j,name in enumerate(npos) if name==pos[i]]) # search for slice/position number for this tile\n",
    "\n",
    "        d=data_dict_organizer(d,'append',\n",
    "                              fov=np.full(len(gene_rol['gene_id'][i]),i),\n",
    "                              gene_rol_id=np.array(gene_rol['gene_id'][i]),\n",
    "                              pos_10x_allx=coord[folders[i]]['lroi10x_x'],\n",
    "                              pos_10x_ally=coord[folders[i]]['lroi10x_y'],\n",
    "                              pos_40x_allx=np.array(gene_rol['lroi_x'][i]),\n",
    "                              pos_40x_ally=np.array(gene_rol['lroi_y'][i]),\n",
    "                              cellidall=np.array(cellid[folders[i]]['cellid'])+np.array(i*starting_fov_idx*dummy_cell_num),# if len(cellid[folders[i]]['cellid']) else np.array([0]),\n",
    "                              sliceidall=np.full(len(gene_rol['gene_id'][i]),pos_id+starting_slice_idx), # check this later,does it require -1 or not\n",
    "                              hyb_rol_id=hyb_rol['gene_id'][i][0],\n",
    "                              fov_hyb=np.full(len(hyb_rol['gene_id'][i][0]),i),\n",
    "                              pos_10x_allx_hyb=coord[folders[i]]['lroi10xhyb_x'],\n",
    "                              pos_10x_ally_hyb=coord[folders[i]]['lroi10xhyb_y'],\n",
    "                              pos_40x_allx_hyb=hyb_rol['lroi_x'][i][0],\n",
    "                              pos_40x_ally_hyb=hyb_rol['lroi_y'][i][0],\n",
    "                              cellidall_hyb=np.array(cellid[folders[i]]['cellidhyb'])+np.array(i*starting_fov_idx*dummy_cell_num),\n",
    "                              sliceidall_hyb=np.full(len(hyb_rol['gene_id'][i][0]),pos_id+starting_slice_idx),\n",
    "                              cell_list_all=np.array(seg[folders[i]]['cell_num'])+np.array(i*starting_fov_idx*dummy_cell_num),\n",
    "                              cell_pos_10x_allx=coord[folders[i]]['cellpos10x_x'],\n",
    "                              cell_pos_10x_ally=coord[folders[i]]['cellpos10x_y'],\n",
    "                              cell_pos_40x_allx=seg[folders[i]]['cent_x'],\n",
    "                              cell_pos_40x_ally=seg[folders[i]]['cent_y'],\n",
    "                              fov_cell=np.full(len(seg[folders[i]]['cell_num']),i),\n",
    "                              sliceidall_cell=np.full(len(seg[folders[i]]['cell_num']),pos_id+starting_slice_idx))\n",
    "\n",
    "    d=data_dict_organizer(d,'concat',fov=[],gene_rol_id=[],pos_10x_allx=[],pos_10x_ally=[],pos_40x_allx=[],pos_40x_ally=[],\n",
    "                          cellidall=[],sliceidall=[],hyb_rol_id=[],fov_hyb=[],pos_10x_allx_hyb=[],pos_10x_ally_hyb=[],pos_40x_allx_hyb=[],pos_40x_ally_hyb=[],\n",
    "                          cellidall_hyb=[],sliceidall_hyb=[],cell_list_all=[],cell_pos_10x_allx=[],cell_pos_10x_ally=[],cell_pos_40x_allx=[],cell_pos_40x_ally=[],\n",
    "                          fov_cell=[],sliceidall_cell=[])\n",
    "    if is_optseq:\n",
    "        print('Has optseq')\n",
    "        codebook_optseq=load(os.path.join(pth,'processed','codebook_optseq.joblib'))\n",
    "        d['hyb_rol_id1']=d['hyb_rol_id']+len(codebook[0])-1+len(codebook_optseq[0])-1\n",
    "        codebook_comb=[codebook[0],codebook_optseq[0],hyb_codebook[0]]\n",
    "    else:\n",
    "        d['hyb_rol_id1']=d['hyb_rol_id']+len(codebook[0])-1\n",
    "        codebook_comb=[codebook[0],hyb_codebook[0]]\n",
    "    codebook_comb=np.concatenate(codebook_comb)\n",
    "    d=merge_gene_hyb_dict(d,'gene_rol_id','hyb_rol_id1','combined_gene_hyb_id')\n",
    "    d=merge_gene_hyb_dict(d,'fov','fov_hyb','combined_gene_hyb_fov')\n",
    "    d=merge_gene_hyb_dict(d,'pos_10x_allx','pos_10x_allx_hyb','combined_gene_hyb_pos10x_x')\n",
    "    d=merge_gene_hyb_dict(d,'pos_10x_ally','pos_10x_ally_hyb','combined_gene_hyb_pos10x_y')\n",
    "    d=merge_gene_hyb_dict(d,'pos_40x_allx','pos_40x_allx_hyb','combined_gene_hyb_pos40x_x')\n",
    "    d=merge_gene_hyb_dict(d,'pos_40x_ally','pos_40x_ally_hyb','combined_gene_hyb_pos40x_y')\n",
    "    d=merge_gene_hyb_dict(d,'cellidall','cellidall_hyb','combined_gene_hyb_cellidall')\n",
    "    d=merge_gene_hyb_dict(d,'sliceidall','sliceidall_hyb','combined_gene_hyb_sliceidall')\n",
    "\n",
    "    \n",
    "    border_size=np.round(fraction_border*tilesize)\n",
    "\n",
    "    pos_id=d['combined_gene_hyb_id']>0 # uncalled rolonies--how does this happen? what's bardensr's code for uncalled ones\n",
    "    pos_inside_border_x=(d['combined_gene_hyb_pos40x_x']>border_size-1) & (d['combined_gene_hyb_pos40x_x']<tilesize-border_size+1)\n",
    "    pos_inside_border_y=(d['combined_gene_hyb_pos40x_y']>border_size-1) & (d['combined_gene_hyb_pos40x_y']<tilesize-border_size+1)\n",
    "    filter_id=pos_id & pos_inside_border_x & pos_inside_border_y\n",
    "\n",
    "    filtered_d={}\n",
    "    filtered_d=data_dict_organizer(filtered_d,'initialize',combined_gene_hyb_id=[],combined_gene_hyb_fov=[],combined_gene_hyb_pos10x_x=[],combined_gene_hyb_pos10x_y=[],\n",
    "                                   combined_gene_hyb_pos40x_x=[],combined_gene_hyb_pos40x_y=[],combined_gene_hyb_cellidall=[],combined_gene_hyb_sliceidall=[])\n",
    "\n",
    "    filtered_d=data_dict_organizer(filtered_d,'append',\n",
    "                                   combined_gene_hyb_id=d['combined_gene_hyb_id'][filter_id],\n",
    "                                   combined_gene_hyb_fov=d['combined_gene_hyb_fov'][filter_id],\n",
    "                                   combined_gene_hyb_pos10x_x=d['combined_gene_hyb_pos10x_x'][filter_id],\n",
    "                                   combined_gene_hyb_pos10x_y=d['combined_gene_hyb_pos10x_y'][filter_id],\n",
    "                                   combined_gene_hyb_pos40x_x=d['combined_gene_hyb_pos40x_x'][filter_id],\n",
    "                                   combined_gene_hyb_pos40x_y=d['combined_gene_hyb_pos40x_y'][filter_id],\n",
    "                                   combined_gene_hyb_cellidall=d['combined_gene_hyb_cellidall'][filter_id],\n",
    "                                   combined_gene_hyb_sliceidall=d['combined_gene_hyb_sliceidall'][filter_id])\n",
    "    filtered_d=data_dict_organizer(filtered_d,'concat',combined_gene_hyb_id=[],combined_gene_hyb_fov=[],combined_gene_hyb_pos10x_x=[],combined_gene_hyb_pos10x_y=[],\n",
    "                                   combined_gene_hyb_pos40x_x=[],combined_gene_hyb_pos40x_y=[],combined_gene_hyb_cellidall=[],combined_gene_hyb_sliceidall=[])\n",
    "\n",
    "    \n",
    "    cells=d['cell_list_all'].copy() # check if copy messed something\n",
    "    genes=np.unique(d['combined_gene_hyb_id'])\n",
    "    rol_id=d['combined_gene_hyb_id'].copy()\n",
    "    rol_cell=d['combined_gene_hyb_cellidall'].copy()\n",
    "    v=pd.crosstab(rol_cell, rol_id, rownames=['cell_index'], colnames=['genes'],dropna=False)\n",
    "    v=v.reindex(index=cells, columns=genes, fill_value=0)\n",
    "    exp_m=coo_matrix(v.to_numpy())\n",
    "    processed_data={'all_data':d,\n",
    "                    'filtered_data':filtered_d,\n",
    "                    'expmat':exp_m,\n",
    "                    'cells':cells,\n",
    "                    'gene_id':genes,\n",
    "                    'codebook_combined':codebook_comb}\n",
    "    dump(processed_data,os.path.join(pth,'processed','processeddata.joblib'))\n",
    "\n",
    "\n",
    "    rolonies={'id':filtered_d['combined_gene_hyb_id'],\n",
    "              'pos10_x':filtered_d['combined_gene_hyb_pos10x_x'],\n",
    "              'pos10_y':filtered_d['combined_gene_hyb_pos10x_y'],\n",
    "              'pos40_x':filtered_d['combined_gene_hyb_pos40x_x'],\n",
    "              'pos40_y':filtered_d['combined_gene_hyb_pos40x_y'],\n",
    "              'slice':filtered_d['combined_gene_hyb_sliceidall'],\n",
    "              'genes':codebook_comb,\n",
    "              'fov':filtered_d['combined_gene_hyb_fov'],\n",
    "              'fov_names':folders}\n",
    "    neurons={'expmat':exp_m,\n",
    "             'id':d['cell_list_all'],\n",
    "             'pos10x_x':d['cell_pos_10x_allx'],\n",
    "             'pos10x_y':d['cell_pos_10x_ally'],\n",
    "             'pos40x_x':d['cell_pos_40x_allx'],\n",
    "             'pos40x_y':d['cell_pos_40x_ally'],\n",
    "             'slice':d['sliceidall_cell'],\n",
    "             'genes':codebook_comb,\n",
    "             'fov':d['fov_cell'],\n",
    "             'fov_names':folders}\n",
    "    dated_filename='alldata'+str(today)+'.joblib'\n",
    "    dump({\"rolonies\":rolonies,\"neurons\":neurons},os.path.join(pth,'processed',dated_filename))\n",
    "    print('ALL DATA IS ORGANIZED')\n",
    "    return dated_filename"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "8644cee2-fad2-40e6-89a4-7fa34994dc67",
   "metadata": {},
   "outputs": [],
   "source": [
    "def data_dict_organizer(d,operation,**kwargs):\n",
    "    \"\"\"\n",
    "    Helper function: Organizes dictionaries\n",
    "    \"\"\"\n",
    "    if operation=='initialize':\n",
    "        d.update(kwargs)\n",
    "    elif operation=='append':\n",
    "        for key in kwargs:\n",
    "            d[key].append(kwargs[key])\n",
    "    elif operation=='concat':\n",
    "        for key in kwargs:\n",
    "            d[key]=np.concatenate(d[key])\n",
    "    return d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "cecfce33-49b6-4e77-81b9-54cfecbc3d72",
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_gene_hyb_dict(d,key1,key2,key3):\n",
    "    \"\"\"\n",
    "    Helper function: Combines gene and hyb data into one dictionary\n",
    "    \"\"\"\n",
    "    ar=[d[key1],d[key2]]\n",
    "    d[key3]=np.concatenate(ar)\n",
    "    return d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "583bfbf1-1503-4923-8310-c1730a87fb38",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.spatial.distance import cdist\n",
    "from scipy.sparse import coo_matrix\n",
    "def filter_overlapping_cells(pth,fname='alldata06092025.joblib',rescale_factor=0.5,px=0.33,box_half_width_um=5,search_radius_um=200):\n",
    "    \"\"\"\n",
    "    Postprocessing function:\n",
    "    1. Removes overlapping neurons per tile\n",
    "    2. Searching within a circle of search radius 200 um around per cell, looking for only cells from different fov and removing cells within a square of overlap width 10um\n",
    "    3. Saves the final output filt_neurons\n",
    "    \"\"\"\n",
    "    data=load(os.path.join(pth,'processed',fname))\n",
    "\n",
    "    neurons=data['neurons']\n",
    "    center_x=neurons['pos10x_x']\n",
    "    center_y=neurons['pos10x_y']\n",
    "    pr=px/rescale_factor\n",
    "    overlap_half_width=np.round(box_half_width_um/pr)\n",
    "    search_radius=np.round(search_radius_um/pr)\n",
    "    xmin=center_x-overlap_half_width\n",
    "    xmax=center_x+overlap_half_width\n",
    "    ymin=center_y-overlap_half_width\n",
    "    ymax=center_y+overlap_half_width\n",
    "    c=np.column_stack((center_x,center_y))\n",
    "\n",
    "    #max_dist=np.sqrt(2)*search_radius\n",
    "    dist=cdist(c,c,'euclidean')\n",
    "    dist_nearest=(dist<search_radius)\n",
    "    [cell_id,nearest_neigh_id]=np.nonzero(dist_nearest)\n",
    "    fov_neigh=neurons['fov'][nearest_neigh_id]\n",
    "    fov_cell=neurons['fov'][cell_id]\n",
    "\n",
    "    sel_cells_id=fov_cell!=fov_neigh\n",
    "    distances_neigh=dist[cell_id[sel_cells_id],nearest_neigh_id[sel_cells_id]]\n",
    "    search_cells_id=cell_id[sel_cells_id]\n",
    "    search_neighbors_id=nearest_neigh_id[sel_cells_id]\n",
    "    search_dist=distances_neigh\n",
    "    id_overlap=((((xmin[search_cells_id]<xmin[search_neighbors_id])&(xmin[search_neighbors_id]<xmax[search_cells_id])) |\n",
    "                 ((xmin[search_cells_id]<xmax[search_neighbors_id])&(xmax[search_neighbors_id]<xmax[search_cells_id]))) & \n",
    "                 (((ymin[search_cells_id]<ymin[search_neighbors_id])&(ymin[search_neighbors_id]<ymax[search_cells_id])) | \n",
    "                 ((ymin[search_cells_id]<ymax[search_neighbors_id])&(ymax[search_neighbors_id]<ymax[search_cells_id]))))\n",
    "    overlap_cells_id=search_cells_id[id_overlap]\n",
    "    overlap_neighbors_id=search_neighbors_id[id_overlap]\n",
    "    overlap_distance=search_dist[id_overlap]\n",
    "    # for i,idc in enumerate(overlap_cells_id):\n",
    "    #     print(f\"Cell {neurons['id'][idc]} in fov {neurons['fov_names'][neurons['fov'][idc]]} is matched to cell {neurons['id'][overlap_neighbors_id[i]]} in fov {neurons['fov_names'][neurons['fov'][overlap_neighbors_id[i]]]} with distance {overlap_distance[i]}\")\n",
    "    exp_mat=neurons['expmat'].todense() # I should do it in csr rather than dense--memory efficient\n",
    "    total_exp_cell=np.asarray(np.sum(exp_mat,axis=1))\n",
    "    [_,rev_idx]=np.unique(neurons['id'][overlap_cells_id],return_counts=False,return_inverse=True)\n",
    "    df=pd.DataFrame({'cell':neurons['id'][overlap_cells_id],'neigh':neurons['id'][overlap_neighbors_id] ,'group':np.transpose(rev_idx),'neigh_exp':total_exp_cell[overlap_neighbors_id].flatten(),'cell_exp':total_exp_cell[overlap_cells_id].flatten()})\n",
    "    idx_max=df.groupby('group')['neigh_exp'].idxmax()   \n",
    "    # df=pd.DataFrame({'cell':neurons['id'][overlap_cells_id],'neigh':neurons['id'][overlap_neighbors_id] ,'neigh_exp':total_exp_cell[overlap_neighbors_id].flatten(),'cell_exp':total_exp_cell[overlap_cells_id].flatten()})\n",
    "    # idx_max=df.groupby('cell')['neigh_exp'].idxmax()   \n",
    "    df_ref=df.loc[idx_max]\n",
    "    remove_cell=df_ref['neigh_exp']>df_ref['cell_exp']\n",
    "    is_removed=np.zeros(center_x.shape)\n",
    "    is_removed[overlap_cells_id[idx_max][remove_cell]]=1\n",
    "    id_to_keep=is_removed==0\n",
    "    filt_neurons={}\n",
    "    expmat=coo_matrix(exp_mat[id_to_keep,:])\n",
    "    filt_neurons['expmat']=expmat\n",
    "    filt_neurons['id']=neurons['id'][id_to_keep]\n",
    "    filt_neurons['pos10x_x']=neurons['pos10x_x'][id_to_keep]\n",
    "    filt_neurons['pos10x_y']=neurons['pos10x_y'][id_to_keep]\n",
    "    filt_neurons['pos40x_x']=neurons['pos40x_x'][id_to_keep]\n",
    "    filt_neurons['pos40x_y']=neurons['pos40x_y'][id_to_keep]\n",
    "    filt_neurons['slice']=neurons['slice'][id_to_keep]\n",
    "    filt_neurons['genes']=neurons['genes']\n",
    "    filt_neurons['fov']=neurons['fov'][id_to_keep]\n",
    "    filt_neurons['fov_names']=neurons['fov_names']\n",
    "    dump({\"filt_neurons\":filt_neurons,\"removecells_all\":id_to_keep},os.path.join(pth,'processed','filt_neurons.joblib'))\n",
    "    print('OVERLAPPING CELLS REMOVED--PROCESSING FINISHED')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "cee41f57-27eb-48e4-b018-1995e77719c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess,yaml\n",
    "def run_diff_env_scripts(name,config_name, **params):\n",
    "    cfg=yaml.safe_load(open(config_name))\n",
    "    spec=cfg[\"tasks\"][name]\n",
    "    argv=[item.format(**params) for item in spec[\"argv\"]]\n",
    "    env=spec[\"env\"].split(\":\",1)[1]\n",
    "    cmd=[\"conda\", \"run\", \"-n\", env, \"python\", *argv]\n",
    "    subprocess.run(cmd, check=True)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "a469ea0b-c39d-420d-9d1b-87a2f1410f54",
   "metadata": {},
   "outputs": [],
   "source": [
    "def preprocessing(pth,config_pth,basedir,chshift_filename_geneseq,chshift_filename_bcseq,chshift_filename_hyb,chprofile_filename_geneseq,chprofile_filename_bcseq,chprofile_filename_hyb,\n",
    "                 config_file='cfg.yaml',is_affine=0,num_initial_c=5,num_later_c=4,num_c=4,subsample_rate=4,do_coarse=0,resize_factor=2,block_size=128,radius=31,local=1,\n",
    "                 cyclename_geneseq='n2vgene',cyclename_bcseq='regn2vbc',cycle_name_hyb='n2vhyb',strength=0.6, m=0,bcidx=4,geneidx=4,ch_bc=4,ch_bc_init=5,bcfname='n2vbcseq',\n",
    "                 num_c_hyb=5,cyclename_rgb='n2vhyb01'):\n",
    "\n",
    "    process_geneseq_cycle(pth,config_pth,basedir,chshift_filename_geneseq,chprofile_filename_geneseq,config_file=config_file,is_affine=is_affine,num_initial_c=num_initial_c,num_later_c=num_later_c,\n",
    "                          num_c=num_c,subsample_rate=subsample_rate,do_coarse=do_coarse,resize_factor=resize_factor, block_size=block_size,radius=radius,local=local,cyclename=cyclename_geneseq,\n",
    "                          strength=strength,m=m)\n",
    "\n",
    "    process_bc_cycle(pth,config_pth,basedir,chshift_filename_bcseq,chprofile_filename_bcseq,config_file=config_file,is_affine=is_affine,num_initial_c=num_initial_c,num_later_c=num_later_c,\n",
    "                     num_c=num_c,subsample_rate=subsample_rate,do_coarse=do_coarse,resize_factor=resize_factor, block_size=block_size,radius=radius,local=local,cyclename=cyclename_bcseq, \n",
    "                     bcidx=bcidx,geneidx=geneidx,ch_bc=ch_bc,ch_bc_init=ch_bc_init,bcfname=bcfname,genefname=cyclename_geneseq,strength=strength,m=m)\n",
    "    process_hyb_cycle(pth, config_pth,basedir,chshift_filename_hyb,chprofile_filename_hyb,config_file=config_file,cycle_name=cycle_name_hyb,num_c=num_c_hyb,is_affine=is_affine,\n",
    "                      cyclename_rgb=cyclename_rgb,strength=strength)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "78dff168-f4f7-4585-8be8-1591507858f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def stitch_and_segment(pth,pth_fiji,fname='n2vgeneseq01',stitching_channel=3,overlap=0.23,numcores=8,is_headless=True,grid_direction='UR',px=0.33,num_c=4,\n",
    "                       intensity_scaling=3,transform_rescale_factor=0.5,display_additional_rescale=0.2,\n",
    "                       model_name='cyto3',outname='cell_mask_cyto3.tif',nuc_ch=5,num_chyb=5,num_cgene=4,\n",
    "                       tilesize=3200,diameter=40,config_file='cfg.yaml',dilation_radius=3, stitch_tool='MIST'):\n",
    "    if stitch_tool=='MIST':\n",
    "        sx,sy=stitch_mist_allpos(pth,pth_fiji,fname,stitching_channel,overlap,numcores,is_headless,grid_direction,px,transform_rescale_factor,num_c)\n",
    "        produce_tform_and_stitched_images(pth,sx,sy,intensity_scaling,tilesize,transform_rescale_factor,display_additional_rescale)\n",
    "    else:\n",
    "        # I DON'T HAVE ASHLAR STICTHING SCRIPT HERE\n",
    "        sx,sy=merge_ashlar_results(pth,transform_rescale_factor,num_c)\n",
    "        produce_tform_and_stitched_images(pth,sx,sy,intensity_scaling,tilesize,transform_rescale_factor,display_additional_rescale)\n",
    "        \n",
    "    \n",
    "    prepare_cellpose_input(pth,nuc_ch,num_chyb,num_cgene,tilesize)\n",
    "    run_diff_env_scripts(\"cellpose_segmentation\",config_file,pth=pth,diameter=diameter,outname=outname,model_name=model_name)\n",
    "    import_cellpose_all_tiles(pth,dilation_radius,outname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "26e0c118-e56a-46be-8c90-eaf92321a8da",
   "metadata": {},
   "outputs": [],
   "source": [
    "def basecalls(pth,config_pth,num_channels=4,codebook_name='codebookM1all.mat',codebook_name2='codebook.joblib',fdrthresh=0.05,trim=160,cropf=0.4,noisefloor_ini=0.01,noisefloor_final=0.05,\n",
    "              prev_codebook_len=0,codebook_opt_name='codebook.mat',no_deconv=1,codebook_hyb_name='codebookhyb.mat',cyclename_hyb='alignedn2vhyb',num_c_hyb=4,thresh=[50,60,40,30],\n",
    "              prominence=[50,10,40,10],all_genes_ch=2,is_optseq=0):\n",
    "    \n",
    "    bardensr_call(pth,config_pth,num_channels,codebook_name,fdrthresh,trim,cropf,noisefloor_ini,noisefloor_final)\n",
    "    import_bardensr_results(pth,'bardensrresult.csv',prev_codebook_len,codebook_name2,is_optseq,codebook_opt_name)\n",
    "    basecall_hyb_all_tiles(pth,cyclename_hyb,config_pth,codebook_hyb_name,thresh,prominence,num_c_hyb,all_genes_ch,no_deconv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "d927280f-094b-46a3-9a19-fb5f2c6d1988",
   "metadata": {},
   "outputs": [],
   "source": [
    "def data_aggregation_and_filtering(pth,config_pth,is_optseq=0,hyb_codebook_name='codebookhyb.mat',starting_slice_idx=1,starting_fov_idx=1,dummy_cell_num=10000,tilesize=3200,fraction_border=0.07,\n",
    "                                   rescale_factor=0.5,px=0.33,box_half_width_um=5,search_radius_um=200,is_bc=0):\n",
    "    aggregate_cell_ids_gene_hyb(pth,is_bc)\n",
    "    transform_downsized_coordinates(pth,is_bc)\n",
    "    fname=organize_processed_data(pth,config_pth,is_optseq,hyb_codebook_name,starting_slice_idx,starting_fov_idx,dummy_cell_num,tilesize,fraction_border)\n",
    "    filter_overlapping_cells(pth,fname,rescale_factor,px,box_half_width_um,search_radius_um)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "dfcea2bf-3c99-4b53-b4a3-35ae5d9f2f89",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_geneseq_cycle_alignment(pth,folder,n_cycles=7):\n",
    "    I=[]\n",
    "\n",
    "    fg, ax=plt.subplots(int(np.round(n_cycles/4)),4)\n",
    "    ax=ax.reshape(-1)\n",
    "    for i in range(n_cycles):\n",
    "        I.append(np.sum(tfl.imread(os.path.join(pth,'processed',folder,'aligned','alignedn2vgeneseq'+str(\"%0.2d\"%(i+1))+'.tif'),key=range(0,4,1)),axis=0))\n",
    "        I1=I[0]\n",
    "        I2=I[i]\n",
    "        ax[i].imshow(make_overlay(I1,I2)[1450:1550,1350:1450,:])\n",
    "    fg.set_figheight(10)\n",
    "    fg.set_figwidth(20)\n",
    "    fg.suptitle('Geneseq to first cycle overlaps', fontsize=16)\n",
    "    plt.tight_layout()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "a5748623-7a32-489a-b23f-6a4194f498d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_bcseq_cycle_alignment(pth,folder,n_cycles):\n",
    "    I=[]\n",
    "\n",
    "    fg, ax=plt.subplots(4,int(np.round(n_cycles/4)))\n",
    "    ax=ax.reshape(-1)\n",
    "\n",
    "    for i in range(n_cycles):\n",
    "        I.append(np.sum(tfl.imread(os.path.join(pth,'processed',folder,'aligned','alignedregn2vbcseq'+str(\"%0.2d\"%(i+1))+'.tif'),key=range(0,4,1)),axis=0))\n",
    "        I1=I[0]\n",
    "        I2=I[i]\n",
    "        ax[i].imshow(make_overlay(I1,I2)[1450:1550,1350:1450,:])\n",
    "    fg.set_figheight(20)\n",
    "    fg.set_figwidth(20)\n",
    "    fg.suptitle('Bcseq to first cycle overlaps', fontsize=16)\n",
    "\n",
    "    plt.tight_layout()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "fe84da43-be65-42a3-85d0-b8a2347ee7af",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_overlay(I1,I2):\n",
    "    I_rgb=np.zeros((3,I1.shape[0],I1.shape[1]))\n",
    "    I_rgb[0,:,:]=np.double(np.divide(I1,np.max(I1,axis=None)))*5\n",
    "    I_rgb[1,:,:]=np.double(np.divide(I2,np.max(I2,axis=None)))*5\n",
    "    Irgb=np.uint8(np.clip(I_rgb*255,0,255))\n",
    "    return np.transpose(Irgb,(1,2,0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "c4ba8093-a762-46e6-831a-d048ae88ef40",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_hyb_basecall(pth,config_pth,fov=9,codebook_hyb_name='codebookhyb.mat'):\n",
    "    codebook_hyb_path=os.path.join(config_pth,codebook_hyb_name)\n",
    "    codebook_hyb=scipy.io.loadmat(codebook_hyb_path)['codebookhyb']\n",
    "    L=load(os.path.join(pth,'processed','genehyb.joblib'))\n",
    "    fg,ax=plt.subplots()\n",
    "    for i in range(len(L['lroi_x'][fov])):\n",
    "        ax.scatter(L['lroi_y'][fov][i], L['lroi_x'][fov][i],s=3)#,c=col[i])\n",
    "        ax.invert_yaxis()\n",
    "    plt.legend([codebook_hyb[0][0],codebook_hyb[1][0],codebook_hyb[2][0]])\n",
    "    fg.suptitle('Hyb basecall', fontsize=16)\n",
    "\n",
    "    plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "095b2ba6-54ee-418d-97b0-dc7ae3a5dd58",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_stitched_output(pth,if_bcseq=1,if_hyb=1,if_geneseq=1,if_mist=1):\n",
    "    fg, ax=plt.subplots(1,3)\n",
    "    fg.suptitle('Stitched', fontsize=16)\n",
    "    reg_folder='checkregistration'\n",
    "    \n",
    "    stitch_output=os.path.join(pth,'processed',reg_folder)\n",
    "    if if_geneseq:\n",
    "        geneseq_pos=tfl.imread(os.path.join(stitch_output,'Pos1_RGBalignedn2vgeneseq01.tif'))\n",
    "        ax[0].imshow(geneseq_pos)\n",
    "    if if_bcseq:\n",
    "        bcseq_pos=tfl.imread(os.path.join(stitch_output,'Pos1_RGBalignedregn2vbcseq01.tif'))\n",
    "        ax[1].imshow(bcseq_pos)\n",
    "    if if_hyb:\n",
    "        hybseq_pos=tfl.imread(os.path.join(stitch_output,'Pos1_RGBalignedn2vhyb01.tif'))\n",
    "        ax[2].imshow(hybseq_pos)\n",
    "    fg.set_figheight(5)\n",
    "    fg.set_figwidth(10)\n",
    "    fg.tight_layout()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "bfc8e82a-a031-45ee-9635-5c14ceb1654f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_final_overlap_removal(pth,data_file):\n",
    "    data2=load(os.path.join(pth,'processed','filt_neurons.joblib'))\n",
    "    data=load(os.path.join(pth,'processed',data_file))\n",
    "    neurons=data['neurons']\n",
    "    filt_neurons=data2['filt_neurons']\n",
    "    id_to_keep=data2['removecells_all']\n",
    "    fg, ax=plt.subplots(1,2)\n",
    "    fg.suptitle('Before and after overlap filter', fontsize=16)\n",
    "    fg.set_figheight(5)\n",
    "    fg.set_figwidth(10)\n",
    "    \n",
    "    exp_mat=neurons['expmat'].todense() # I should do it in csr rather than dense--memory efficient\n",
    "    total_exp_cell=np.asarray(np.sum(exp_mat,axis=1))\n",
    "    ax[0].scatter(neurons['pos10x_x'],neurons['pos10x_y'],1,total_exp_cell,cmap=plt.cm.viridis_r)\n",
    "    ax[1].scatter(filt_neurons['pos10x_x'],filt_neurons['pos10x_y'],1,total_exp_cell[id_to_keep],cmap=plt.cm.viridis_r)\n",
    "    fg.tight_layout()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "51be1db6-cd7d-4907-bae4-d105de63c3b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_cellpose_output(pth,folder):\n",
    "    I=np.sum(tfl.imread(os.path.join(pth,'processed',folder,'aligned','cell_inp2.tif'),key=range(0,2)),axis=0)\n",
    "    mask=tfl.imread(os.path.join(pth,'processed',folder,'aligned','cell_mask_cyto3.tif'))\n",
    "    mask_dil=expand_labels(mask)\n",
    "    fg, ax=plt.subplots(1,2)\n",
    "    fg.suptitle('Cell labels', fontsize=16)\n",
    "    ax[0].imshow(I)\n",
    "    ax[1].imshow(mask_dil)\n",
    "    plt.tight_layout()\n",
    "    fg.set_figheight(10)\n",
    "    fg.set_figwidth(20)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "b93ab147-bfd3-48f4-b058-8d58e5c9e674",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_gene_expression(pth,gene_idx):\n",
    "    fig = plt.figure()\n",
    "    data2=load(os.path.join(pth,'processed','filt_neurons.joblib'))\n",
    "    filt_neurons=data2['filt_neurons']\n",
    "    genes=np.unique(filt_neurons['id'])\n",
    "    exp_m=filt_neurons['expmat']\n",
    "    idg=exp_m.col==gene_idx\n",
    "    plt.scatter(filt_neurons['pos10x_x'][exp_m.row[idg]],filt_neurons['pos10x_y'][exp_m.row[idg]],2,np.log(1+exp_m.data[idg]),cmap='cool')\n",
    "    #plt.savefig(os.path.join(pth,'processed','png',str(codebook_comb[gene_idx][0])+'.png'),dpi=300)\n",
    "    fig.suptitle(f\"{filt_neurons['genes'][gene_idx][0]} expression\", fontsize=16)\n",
    "\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "1c45413e-237e-42fd-8ebb-0c861180a560",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_cycle_registration(pth,thresh,modality,n_cycles=7):\n",
    "    [folders,pos,_,_]=get_folders(pth)\n",
    "    N=len(folders)\n",
    "    v=np.zeros([N,n_cycles])\n",
    "    cont=np.zeros([N,1])\n",
    "    #q=random.randint(1, len(folders))\n",
    "    t0=time.time()\n",
    "    for q in range(N):\n",
    "        print(f\"folder {folders[q]}\")\n",
    "        for i in range(n_cycles):\n",
    "            I2=np.sum(tfl.imread(os.path.join(pth,'processed',folders[q],'aligned','alignedfixedn2vgeneseq'+str(\"%0.2d\"%(i+1))+'.tif'),key=range(0,4,1)),axis=0)\n",
    "            if i==0:\n",
    "                I1=I2\n",
    "            v[q,i]=nmi(I1,I2)\n",
    "        cont[q,0]=np.var(I1)\n",
    "\n",
    "    variance_fov=np.zeros([v.shape[0],1])\n",
    "    eps=0\n",
    "    for r in range(v.shape[0]):\n",
    "        vr=((v[r,1:]-np.min(v[r,1:]))/(eps+np.max(v[r,1:])-np.min(v[r,1:])))\n",
    "        variance_fov[r]=np.max(vr)/(eps+np.median(vr))\n",
    "    \n",
    "    idx=np.abs(variance_fov)>thresh\n",
    "    idx=idx.reshape(-1)\n",
    "    folders=np.array(folders)\n",
    "    sel_folders=folders[idx]\n",
    "    sel_folders_score=variance_fov[idx]\n",
    "    sel_folders_cont=cont[idx]\n",
    "    tn=time.time()\n",
    "    dump({'name':sel_folders,'score':sel_folders_score,'content':sel_folders_cont},os.path.join(pth,'processed',modality+'cycle_registration_metric.joblib'))\n",
    "\n",
    "\n",
    "    print(f\"Total time {tn-t0} seconds\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "9c6fb586-1d16-425f-b50e-0af27b05d53a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def stitch_images_RGB_cycles(pth,intensity_scaling=3,tilesize=3200,transform_rescale_factor=0.5,display_additional_rescale=1):\n",
    "    [folders,pos,x,y]=get_folders(pth)\n",
    "    unique_pos=nsort(np.unique(pos))\n",
    "    check_reg_folder=os.path.join(pth,'processed','checkregistration_rgb_cycles')\n",
    "    os.makedirs(check_reg_folder,exist_ok=True)\n",
    "    folder_names=np.array(folders)\n",
    "    filenames=sorted(glob.glob(os.path.join(pth,'processed','gene_cycle_pngs',\"*.png\")))\n",
    "    T=load(os.path.join(pth,'tforms_rescaled'+str(transform_rescale_factor).replace('.','p')+'.joblib'))\n",
    "    sx=[]\n",
    "    sy=[]\n",
    "    for n_pos in unique_pos:\n",
    "        pos_id=np.array([i for i,name in enumerate(pos) if name==n_pos])\n",
    "        for ids in pos_id:\n",
    "            tilename=folder_names[ids]+'.tif'\n",
    "            sx.append(T[n_pos][tilename][\"ref_pos\"][0])\n",
    "            sy.append(T[n_pos][tilename][\"ref_pos\"][1])\n",
    "   \n",
    "\n",
    "    for n_pos in unique_pos:\n",
    "        I=np.zeros([int(np.max(sy)+tilesize*transform_rescale_factor),int(np.max(sx)+tilesize*transform_rescale_factor),3])\n",
    "        pos_id=np.array([i for i,name in enumerate(pos) if name==n_pos])\n",
    "        for ids in pos_id:\n",
    "            tform=load(os.path.join(pth,'processed',folder_names[ids],'global_tform_'+str(transform_rescale_factor).replace('.','p')+'.joblib'))\n",
    "            tile=tfl.imread(filenames[ids])\n",
    "            It=skimage.transform.warp(np.squeeze(tile),tform.inverse,preserve_range=True,output_shape=(I.shape[0],I.shape[1]))\n",
    "            I=np.maximum(I,It)\n",
    "        Irgb=np.uint8(np.clip(I*intensity_scaling,0,255))\n",
    "        Irgb_rescaled = np.uint8(np.clip(rescale(Irgb, display_additional_rescale, channel_axis=-1, preserve_range=True, anti_aliasing=True),0,255))\n",
    "        tfl.imwrite(os.path.join(check_reg_folder,n_pos+'.png'),Irgb_rescaled,photometric='rgb')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "665409bd-8c3b-4b1c-b877-3edb14b903de",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying files in the correct folders...\n",
      "Organizing folder /mnt/d/allen_data/test_dataset_bc_726126/geneseq01\n",
      "\n",
      "Copied 0\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[75], line 21\u001b[0m\n\u001b[1;32m     16\u001b[0m chshift_filename_hyb\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mchshift20x-hyb-20220218.mat\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m     17\u001b[0m chprofile_filename_hyb\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mchprofilehyb_allen.mat\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m---> 21\u001b[0m \u001b[43mpreprocessing\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpth\u001b[49m\u001b[43m,\u001b[49m\u001b[43mconfig_pth\u001b[49m\u001b[43m,\u001b[49m\u001b[43mbasedir\u001b[49m\u001b[43m,\u001b[49m\u001b[43mchshift_filename_geneseq\u001b[49m\u001b[43m,\u001b[49m\u001b[43mchshift_filename_bcseq\u001b[49m\u001b[43m,\u001b[49m\u001b[43mchshift_filename_hyb\u001b[49m\u001b[43m,\u001b[49m\u001b[43mchprofile_filename_geneseq\u001b[49m\u001b[43m,\u001b[49m\u001b[43mchprofile_filename_bcseq\u001b[49m\u001b[43m,\u001b[49m\u001b[43mchprofile_filename_hyb\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     22\u001b[0m \u001b[43m                 \u001b[49m\u001b[43mconfig_file\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mcfg.yaml\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43mis_affine\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43mnum_initial_c\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m5\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43mnum_later_c\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m4\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43mnum_c\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m4\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43msubsample_rate\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m4\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43mdo_coarse\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43mresize_factor\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m2\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43mblock_size\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m128\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43mradius\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m31\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43mlocal\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     23\u001b[0m \u001b[43m                 \u001b[49m\u001b[43mcyclename_geneseq\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mn2vgene\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43mcyclename_bcseq\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mregn2vbc\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43mcycle_name_hyb\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mn2vhyb\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43mstrength\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m0.6\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mm\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43mbcidx\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m4\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43mgeneidx\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m4\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43mch_bc\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m4\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43mch_bc_init\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m5\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43mbcfname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mn2vbcseq\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     24\u001b[0m \u001b[43m                 \u001b[49m\u001b[43mnum_c_hyb\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m5\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43mcyclename_rgb\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mn2vhyb01\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m     25\u001b[0m stitch_and_segment(pth,pth_fiji,fname\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mn2vgeneseq01\u001b[39m\u001b[38;5;124m'\u001b[39m,stitching_channel\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m3\u001b[39m,overlap\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0.23\u001b[39m,numcores\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m8\u001b[39m,is_headless\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m,grid_direction\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mUR\u001b[39m\u001b[38;5;124m'\u001b[39m,px\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0.33\u001b[39m,num_c\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m4\u001b[39m,\n\u001b[1;32m     26\u001b[0m                        intensity_scaling\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m3\u001b[39m,transform_rescale_factor\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0.5\u001b[39m,display_additional_rescale\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0.2\u001b[39m,\n\u001b[1;32m     27\u001b[0m                        model_name\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcyto3\u001b[39m\u001b[38;5;124m'\u001b[39m,outname\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcell_mask_cyto3.tif\u001b[39m\u001b[38;5;124m'\u001b[39m,nuc_ch\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m5\u001b[39m,num_chyb\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m5\u001b[39m,num_cgene\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m4\u001b[39m,\n\u001b[1;32m     28\u001b[0m                        tilesize\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m3200\u001b[39m,diameter\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m40\u001b[39m,config_file\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcfg.yaml\u001b[39m\u001b[38;5;124m'\u001b[39m,dilation_radius\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m3\u001b[39m)\n\u001b[1;32m     29\u001b[0m basecalls(pth,config_pth,num_channels\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m4\u001b[39m,codebook_name\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcodebookM1all.mat\u001b[39m\u001b[38;5;124m'\u001b[39m,codebook_name2\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcodebook.joblib\u001b[39m\u001b[38;5;124m'\u001b[39m,fdrthresh\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0.05\u001b[39m,trim\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m160\u001b[39m,cropf\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0.4\u001b[39m,noisefloor_ini\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0.01\u001b[39m,noisefloor_final\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0.05\u001b[39m,\n\u001b[1;32m     30\u001b[0m               prev_codebook_len\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m,codebook_opt_name\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcodebook.mat\u001b[39m\u001b[38;5;124m'\u001b[39m,no_deconv\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m,codebook_hyb_name\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcodebookhyb.mat\u001b[39m\u001b[38;5;124m'\u001b[39m,cyclename_hyb\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124malignedn2vhyb\u001b[39m\u001b[38;5;124m'\u001b[39m,num_c_hyb\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m4\u001b[39m,thresh\u001b[38;5;241m=\u001b[39m[\u001b[38;5;241m50\u001b[39m,\u001b[38;5;241m60\u001b[39m,\u001b[38;5;241m40\u001b[39m,\u001b[38;5;241m30\u001b[39m],\n\u001b[1;32m     31\u001b[0m               prominence\u001b[38;5;241m=\u001b[39m[\u001b[38;5;241m50\u001b[39m,\u001b[38;5;241m10\u001b[39m,\u001b[38;5;241m40\u001b[39m,\u001b[38;5;241m10\u001b[39m],all_genes_ch\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m2\u001b[39m,is_optseq\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m)\n",
      "Cell \u001b[0;32mIn[60], line 6\u001b[0m, in \u001b[0;36mpreprocessing\u001b[0;34m(pth, config_pth, basedir, chshift_filename_geneseq, chshift_filename_bcseq, chshift_filename_hyb, chprofile_filename_geneseq, chprofile_filename_bcseq, chprofile_filename_hyb, config_file, is_affine, num_initial_c, num_later_c, num_c, subsample_rate, do_coarse, resize_factor, block_size, radius, local, cyclename_geneseq, cyclename_bcseq, cycle_name_hyb, strength, m, bcidx, geneidx, ch_bc, ch_bc_init, bcfname, num_c_hyb, cyclename_rgb)\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mpreprocessing\u001b[39m(pth,config_pth,basedir,chshift_filename_geneseq,chshift_filename_bcseq,chshift_filename_hyb,chprofile_filename_geneseq,chprofile_filename_bcseq,chprofile_filename_hyb,\n\u001b[1;32m      2\u001b[0m                  config_file\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcfg.yaml\u001b[39m\u001b[38;5;124m'\u001b[39m,is_affine\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m,num_initial_c\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m5\u001b[39m,num_later_c\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m4\u001b[39m,num_c\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m4\u001b[39m,subsample_rate\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m4\u001b[39m,do_coarse\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m,resize_factor\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m2\u001b[39m,block_size\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m128\u001b[39m,radius\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m31\u001b[39m,local\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m,\n\u001b[1;32m      3\u001b[0m                  cyclename_geneseq\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mn2vgene\u001b[39m\u001b[38;5;124m'\u001b[39m,cyclename_bcseq\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mregn2vbc\u001b[39m\u001b[38;5;124m'\u001b[39m,cycle_name_hyb\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mn2vhyb\u001b[39m\u001b[38;5;124m'\u001b[39m,strength\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0.6\u001b[39m, m\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m,bcidx\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m4\u001b[39m,geneidx\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m4\u001b[39m,ch_bc\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m4\u001b[39m,ch_bc_init\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m5\u001b[39m,bcfname\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mn2vbcseq\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m      4\u001b[0m                  num_c_hyb\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m5\u001b[39m,cyclename_rgb\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mn2vhyb01\u001b[39m\u001b[38;5;124m'\u001b[39m):\n\u001b[0;32m----> 6\u001b[0m     \u001b[43mprocess_geneseq_cycle\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpth\u001b[49m\u001b[43m,\u001b[49m\u001b[43mconfig_pth\u001b[49m\u001b[43m,\u001b[49m\u001b[43mbasedir\u001b[49m\u001b[43m,\u001b[49m\u001b[43mchshift_filename_geneseq\u001b[49m\u001b[43m,\u001b[49m\u001b[43mchprofile_filename_geneseq\u001b[49m\u001b[43m,\u001b[49m\u001b[43mconfig_file\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconfig_file\u001b[49m\u001b[43m,\u001b[49m\u001b[43mis_affine\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mis_affine\u001b[49m\u001b[43m,\u001b[49m\u001b[43mnum_initial_c\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mnum_initial_c\u001b[49m\u001b[43m,\u001b[49m\u001b[43mnum_later_c\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mnum_later_c\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m      7\u001b[0m \u001b[43m                          \u001b[49m\u001b[43mnum_c\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mnum_c\u001b[49m\u001b[43m,\u001b[49m\u001b[43msubsample_rate\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msubsample_rate\u001b[49m\u001b[43m,\u001b[49m\u001b[43mdo_coarse\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdo_coarse\u001b[49m\u001b[43m,\u001b[49m\u001b[43mresize_factor\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mresize_factor\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mblock_size\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mblock_size\u001b[49m\u001b[43m,\u001b[49m\u001b[43mradius\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mradius\u001b[49m\u001b[43m,\u001b[49m\u001b[43mlocal\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlocal\u001b[49m\u001b[43m,\u001b[49m\u001b[43mcyclename\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcyclename_geneseq\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m      8\u001b[0m \u001b[43m                          \u001b[49m\u001b[43mstrength\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstrength\u001b[49m\u001b[43m,\u001b[49m\u001b[43mm\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mm\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     10\u001b[0m     process_bc_cycle(pth,config_pth,basedir,chshift_filename_bcseq,chprofile_filename_bcseq,config_file\u001b[38;5;241m=\u001b[39mconfig_file,is_affine\u001b[38;5;241m=\u001b[39mis_affine,num_initial_c\u001b[38;5;241m=\u001b[39mnum_initial_c,num_later_c\u001b[38;5;241m=\u001b[39mnum_later_c,\n\u001b[1;32m     11\u001b[0m                      num_c\u001b[38;5;241m=\u001b[39mnum_c,subsample_rate\u001b[38;5;241m=\u001b[39msubsample_rate,do_coarse\u001b[38;5;241m=\u001b[39mdo_coarse,resize_factor\u001b[38;5;241m=\u001b[39mresize_factor, block_size\u001b[38;5;241m=\u001b[39mblock_size,radius\u001b[38;5;241m=\u001b[39mradius,local\u001b[38;5;241m=\u001b[39mlocal,cyclename\u001b[38;5;241m=\u001b[39mcyclename_bcseq, \n\u001b[1;32m     12\u001b[0m                      bcidx\u001b[38;5;241m=\u001b[39mbcidx,geneidx\u001b[38;5;241m=\u001b[39mgeneidx,ch_bc\u001b[38;5;241m=\u001b[39mch_bc,ch_bc_init\u001b[38;5;241m=\u001b[39mch_bc_init,bcfname\u001b[38;5;241m=\u001b[39mbcfname,genefname\u001b[38;5;241m=\u001b[39mcyclename_geneseq,strength\u001b[38;5;241m=\u001b[39mstrength,m\u001b[38;5;241m=\u001b[39mm)\n\u001b[1;32m     13\u001b[0m     process_hyb_cycle(pth, config_pth,basedir,chshift_filename_hyb,chprofile_filename_hyb,config_file\u001b[38;5;241m=\u001b[39mconfig_file,cycle_name\u001b[38;5;241m=\u001b[39mcycle_name_hyb,num_c\u001b[38;5;241m=\u001b[39mnum_c_hyb,is_affine\u001b[38;5;241m=\u001b[39mis_affine,\n\u001b[1;32m     14\u001b[0m                       cyclename_rgb\u001b[38;5;241m=\u001b[39mcyclename_rgb,strength\u001b[38;5;241m=\u001b[39mstrength)\n",
      "Cell \u001b[0;32mIn[21], line 7\u001b[0m, in \u001b[0;36mprocess_geneseq_cycle\u001b[0;34m(pthb, config_pth, basedir, chshift_filename, chprofile_filename, config_file, is_affine, num_initial_c, num_later_c, num_c, subsample_rate, do_coarse, resize_factor, block_size, radius, local, cyclename, strength, m)\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;124;03mPreprocessing function:\u001b[39;00m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;124;03mFor bcseq--calls functions that align bcseq0 to geneseq0, preprocess bcseq--for all tiles\u001b[39;00m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m   \n\u001b[1;32m      6\u001b[0m t0\u001b[38;5;241m=\u001b[39mtime\u001b[38;5;241m.\u001b[39mtime()\n\u001b[0;32m----> 7\u001b[0m \u001b[43morganize_files\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpthb\u001b[49m\u001b[43m,\u001b[49m\u001b[43mcyclename\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mgeneseq\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      8\u001b[0m denoise_n2v_call(pthb,basedir,fname\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgeneseq\u001b[39m\u001b[38;5;124m'\u001b[39m,config_file\u001b[38;5;241m=\u001b[39mconfig_file)\n\u001b[1;32m      9\u001b[0m register_seq_images_subsample(pthb,config_pth,chshift_filename,chprofile_filename,cyclename,is_affine,num_initial_c, num_later_c,num_c,subsample_rate,do_coarse,resize_factor, block_size,radius,local,strength,m)\n",
      "Cell \u001b[0;32mIn[11], line 25\u001b[0m, in \u001b[0;36morganize_files\u001b[0;34m(pth, cyclename)\u001b[0m\n\u001b[1;32m     23\u001b[0m fname\u001b[38;5;241m=\u001b[39mname[\u001b[38;5;241m1\u001b[39m]\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m.\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m     24\u001b[0m os\u001b[38;5;241m.\u001b[39mmakedirs(os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(pth,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mprocessed\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mMAX\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m+\u001b[39mfname[\u001b[38;5;241m0\u001b[39m]),exist_ok\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[0;32m---> 25\u001b[0m \u001b[43mshutil\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcopy\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf1\u001b[49m\u001b[43m,\u001b[49m\u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpath\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mjoin\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpth\u001b[49m\u001b[43m,\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mprocessed\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mMAX\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43mfname\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43mcyclename\u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43mf2\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpartition\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcyclename\u001b[49m\u001b[43m)\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m-\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m.tif\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     26\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mCopied \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mj\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m)\n",
      "File \u001b[0;32m~/anaconda3/envs/barseq/lib/python3.8/shutil.py:418\u001b[0m, in \u001b[0;36mcopy\u001b[0;34m(src, dst, follow_symlinks)\u001b[0m\n\u001b[1;32m    416\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39misdir(dst):\n\u001b[1;32m    417\u001b[0m     dst \u001b[38;5;241m=\u001b[39m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(dst, os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mbasename(src))\n\u001b[0;32m--> 418\u001b[0m \u001b[43mcopyfile\u001b[49m\u001b[43m(\u001b[49m\u001b[43msrc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdst\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfollow_symlinks\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfollow_symlinks\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    419\u001b[0m copymode(src, dst, follow_symlinks\u001b[38;5;241m=\u001b[39mfollow_symlinks)\n\u001b[1;32m    420\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m dst\n",
      "File \u001b[0;32m~/anaconda3/envs/barseq/lib/python3.8/shutil.py:275\u001b[0m, in \u001b[0;36mcopyfile\u001b[0;34m(src, dst, follow_symlinks)\u001b[0m\n\u001b[1;32m    273\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m _USE_CP_SENDFILE:\n\u001b[1;32m    274\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 275\u001b[0m         \u001b[43m_fastcopy_sendfile\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfsrc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfdst\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    276\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m dst\n\u001b[1;32m    277\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m _GiveupOnFastCopy:\n",
      "File \u001b[0;32m~/anaconda3/envs/barseq/lib/python3.8/shutil.py:152\u001b[0m, in \u001b[0;36m_fastcopy_sendfile\u001b[0;34m(fsrc, fdst)\u001b[0m\n\u001b[1;32m    150\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[1;32m    151\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 152\u001b[0m         sent \u001b[38;5;241m=\u001b[39m \u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msendfile\u001b[49m\u001b[43m(\u001b[49m\u001b[43moutfd\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43minfd\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43moffset\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mblocksize\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    153\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mOSError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[1;32m    154\u001b[0m         \u001b[38;5;66;03m# ...in oder to have a more informative exception.\u001b[39;00m\n\u001b[1;32m    155\u001b[0m         err\u001b[38;5;241m.\u001b[39mfilename \u001b[38;5;241m=\u001b[39m fsrc\u001b[38;5;241m.\u001b[39mname\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# NEED TO ADD SOME CHECKPOINTS SO THAT WHATEVER IS DONE IS NOT REPEATED\n",
    "# NEED TO ADD CHECKPOINTS TO SEE IF DIFFERENT MODALITIES EXIST IN THIS EXPERIMENT\n",
    "\n",
    "pth='/mnt/d/allen_data/test_dataset_bc_726126/'\n",
    "pth_fiji = r'/home/nrg/Fiji.app' \n",
    "config_pth='/mnt/c/Users/goswami/Documents/MATLAB/barseq-matlab-allen-main (1)/barseq-matlab-allen-main/config/'\n",
    "basedir = '/home/nrg/n2vmodels/'\n",
    "\n",
    "\n",
    "chshift_filename_geneseq='chshift20x-20220218.mat'\n",
    "chprofile_filename_geneseq='chprofile20x-50-30-20-40-20220218.mat'\n",
    "\n",
    "chshift_filename_bcseq='chshift20x-20220218.mat'\n",
    "chprofile_filename_bcseq='chprofile20x-50-30-20-40-20220218.mat'\n",
    "\n",
    "chshift_filename_hyb='chshift20x-hyb-20220218.mat'\n",
    "chprofile_filename_hyb='chprofilehyb_allen.mat'\n",
    "\n",
    "\n",
    "\n",
    "preprocessing(pth,config_pth,basedir,chshift_filename_geneseq,chshift_filename_bcseq,chshift_filename_hyb,chprofile_filename_geneseq,chprofile_filename_bcseq,chprofile_filename_hyb,\n",
    "                 config_file='cfg.yaml',is_affine=0,num_initial_c=5,num_later_c=4,num_c=4,subsample_rate=4,do_coarse=0,resize_factor=2,block_size=128,radius=31,local=1,\n",
    "                 cyclename_geneseq='n2vgene',cyclename_bcseq='regn2vbc',cycle_name_hyb='n2vhyb',strength=0.6, m=0,bcidx=4,geneidx=4,ch_bc=4,ch_bc_init=5,bcfname='n2vbcseq',\n",
    "                 num_c_hyb=5,cyclename_rgb='n2vhyb01')\n",
    "stitch_and_segment(pth,pth_fiji,fname='n2vgeneseq01',stitching_channel=3,overlap=0.23,numcores=8,is_headless=True,grid_direction='UR',px=0.33,num_c=4,\n",
    "                       intensity_scaling=3,transform_rescale_factor=0.5,display_additional_rescale=0.2,\n",
    "                       model_name='cyto3',outname='cell_mask_cyto3.tif',nuc_ch=5,num_chyb=5,num_cgene=4,\n",
    "                       tilesize=3200,diameter=40,config_file='cfg.yaml',dilation_radius=3)\n",
    "basecalls(pth,config_pth,num_channels=4,codebook_name='codebookM1all.mat',codebook_name2='codebook.joblib',fdrthresh=0.05,trim=160,cropf=0.4,noisefloor_ini=0.01,noisefloor_final=0.05,\n",
    "              prev_codebook_len=0,codebook_opt_name='codebook.mat',no_deconv=1,codebook_hyb_name='codebookhyb.mat',cyclename_hyb='alignedn2vhyb',num_c_hyb=4,thresh=[50,60,40,30],\n",
    "              prominence=[50,10,40,10],all_genes_ch=2,is_optseq=0)\n",
    "data_aggregation_and_filtering(pth,config_pth,is_optseq=0,hyb_codebook_name='codebookhyb.mat',starting_slice_idx=1,starting_fov_idx=1,dummy_cell_num=10000,tilesize=3200,fraction_border=0.07,\n",
    "                                   rescale_factor=0.5,px=0.33,box_half_width_um=5,search_radius_um=200,is_bc=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6daf89b5-2970-4873-a362-be37b0c7a47d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# RUN SOME TESTS\n",
    "data_file='alldata19092025.joblib'\n",
    "gene_idx=111\n",
    "[folders,pos,_,_]=get_folders(pth)\n",
    "q=9\n",
    "folder=folders[q]\n",
    "check_geneseq_cycle_alignment(pth,folder,7)\n",
    "check_bcseq_cycle_alignment(pth,folder,15)\n",
    "check_hyb_basecall(pth,config_pth,fov=q,codebook_hyb_name='codebookhyb.mat')\n",
    "check_stitched_output(pth,if_bcseq=1,if_hyb=1,if_geneseq=1,if_mist=1)\n",
    "check_cellpose_output(pth,folder)\n",
    "check_final_overlap_removal(pth,data_file)\n",
    "check_gene_expression(pth,gene_idx)\n",
    "stitch_images_RGB_cycles(pth,intensity_scaling=3,tilesize=3200,transform_rescale_factor=0.5,display_additional_rescale=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "989e9d77-cfca-4deb-b031-f03a69b81406",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "pth='/mnt/d/allen_data/test_dataset_bc_726126/processed/MAX_Pos1_000_003/original/'\n",
    "\n",
    "config_pth='/mnt/c/Users/goswami/Documents/MATLAB/barseq-matlab-allen-main (1)/barseq-matlab-allen-main/config/'\n",
    "cyclename='n2vgene'\n",
    "is_affine=0\n",
    "num_initial_c=5 # initial channels in first round of sequencing to read\n",
    "num_later_c=4 # subsequent cycle channels to read\n",
    "num_c=4 # channels to perform bleedthrough-correction and channel alignment, needs more elegant solution--ng\n",
    "subsample_rate=4\n",
    "do_coarse=0\n",
    "resize_factor=2\n",
    "block_size=128\n",
    "chshift_filename='chshift20x-20220218.mat'\n",
    "chprofile_filename='chprofile20x-50-30-20-40-20220218.mat'\n",
    "radius=31\n",
    "local=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7de034e-8478-41fc-b393-b4cbb368a2f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "I=tfl.imread(os.path.join(pth,'n2vgeneseq02.tif'),key=range(0,4,1))\n",
    "Ifilt=back_sub_opencv_open(I,31,pth,'bcksb.tif',4,0)\n",
    "\n",
    "fg, ax=plt.subplots(1,2)\n",
    "ax[0].imshow(I[3,:,:],cmap='jet',vmin=0,vmax=150)\n",
    "ax[1].imshow(Ifilt[3,:,:],cmap='jet',vmin=0,vmax=150)\n",
    "plt.tight_layout()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fd3e9a7-e2e6-462e-8909-003bc51a392d",
   "metadata": {},
   "outputs": [],
   "source": [
    "Ialigned=channel_alignment(Ifilt,chshift_filename,config_pth,pth,'aligned.tif',4,0,0)\n",
    "fg, ax=plt.subplots(1,2)\n",
    "ax[0].imshow(np.sum(Ifilt[:,362:372,450:462],axis=0),cmap='jet',vmin=0,vmax=150)\n",
    "ax[1].imshow(np.sum(Ialigned[:,362:372,450:462],axis=0),cmap='jet',vmin=0,vmax=150)\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93f68bf8-efae-4aea-acc1-9eaedc24fff8",
   "metadata": {},
   "outputs": [],
   "source": [
    "Icorr=bleedthrough_linear(Ialigned,4,config_pth,chprofile_filename,pth,'corr.tif',0)\n",
    "fg, ax=plt.subplots(2,4)\n",
    "for i in range(4):\n",
    "    ax[0,i].imshow(Ialigned[i,362:372,450:462],cmap='jet',vmin=0,vmax=100)\n",
    "    ax[1,i].imshow(Icorr[i,362:372,450:462],cmap='jet',vmin=0,vmax=100)\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab22c7f6-c529-47f5-b0e4-b8613a741ee8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#CHECK GENESEQ CYCLE ALIGNEMENT\n",
    "pth='/mnt/d/allen_data/test_dataset_bc_726126/processed/MAX_Pos1_000_003/aligned/'\n",
    "I1=np.sum(tfl.imread(os.path.join(pth,'alignedn2vgeneseq01.tif'),key=range(0,4,1)),axis=0)\n",
    "I2=np.sum(tfl.imread(os.path.join(pth,'alignedn2vgeneseq07.tif'),key=range(0,4,1)),axis=0)\n",
    "I_rgb=np.zeros((3,I1.shape[0],I1.shape[1]))\n",
    "I_rgb[0,:,:]=np.double(np.divide(I1,np.max(I1,axis=None)))*10\n",
    "I_rgb[1,:,:]=np.double(np.divide(I2,np.max(I2,axis=None)))*10\n",
    "\n",
    "\n",
    "Irgb=np.uint8(np.clip(I_rgb*255,0,255))\n",
    "\n",
    "pthb='/mnt/d/allen_data/test_dataset_bc_726126/726126_test_dataset_processed/processed/MAX_Pos1_000_003/aligned/MAX_Pos1_000_003/aligned'\n",
    "I1b=np.sum(tfl.imread(os.path.join(pthb,'alignedfixedn2vgeneseq01.tif'),key=range(0,4,1)),axis=0)\n",
    "I2b=np.sum(tfl.imread(os.path.join(pthb,'alignedfixedn2vgeneseq07.tif'),key=range(0,4,1)),axis=0)\n",
    "I_rgbb=np.zeros((3,I1b.shape[0],I1b.shape[1]))\n",
    "I_rgbb[0,:,:]=np.double(np.divide(I1b,np.max(I1b,axis=None)))*10\n",
    "I_rgbb[1,:,:]=np.double(np.divide(I2b,np.max(I2b,axis=None)))*10\n",
    "\n",
    "\n",
    "Irgbb=np.uint8(np.clip(I_rgbb*255,0,255))\n",
    "\n",
    "\n",
    "fg, ax=plt.subplots(1,2)\n",
    "ax[0].imshow(np.transpose(Irgb[:,1450:1550,1350:1450],(1,2,0)),vmin=0,vmax=1)\n",
    "ax[1].imshow(np.transpose(Irgbb[:,1450:1550,1350:1450],(1,2,0)),vmin=0,vmax=1)\n",
    "plt.tight_layout()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30b788c9-1b9e-407f-b259-88f485fd2022",
   "metadata": {},
   "outputs": [],
   "source": [
    "#CHECK BCSEQ CYCLE ALIGNEMENT\n",
    "pth='/mnt/d/allen_data/test_dataset_bc_726126/processed/MAX_Pos1_000_003/aligned/'\n",
    "I1=np.sum(tfl.imread(os.path.join(pth,'alignedregn2vbcseq01.tif'),key=range(0,4,1)),axis=0)\n",
    "I2=np.sum(tfl.imread(os.path.join(pth,'alignedregn2vbcseq05.tif'),key=range(0,4,1)),axis=0)\n",
    "I_rgb=np.zeros((3,I1.shape[0],I1.shape[1]))\n",
    "I_rgb[0,:,:]=np.double(np.divide(I1,np.max(I1,axis=None)))*10\n",
    "I_rgb[1,:,:]=np.double(np.divide(I2,np.max(I2,axis=None)))*10\n",
    "\n",
    "\n",
    "Irgb=np.uint8(np.clip(I_rgb*255,0,255))\n",
    "\n",
    "pthb='/mnt/d/allen_data/test_dataset_bc_726126/726126_test_dataset_processed/processed/MAX_Pos1_000_003/aligned/MAX_Pos1_000_003/aligned'\n",
    "I1b=np.sum(tfl.imread(os.path.join(pthb,'alignedfixedregn2vbcseq01.tif'),key=range(0,4,1)),axis=0)\n",
    "I2b=np.sum(tfl.imread(os.path.join(pthb,'alignedfixedregn2vbcseq05.tif'),key=range(0,4,1)),axis=0)\n",
    "I_rgbb=np.zeros((3,I1b.shape[0],I1b.shape[1]))\n",
    "I_rgbb[0,:,:]=np.double(np.divide(I1b,np.max(I1b,axis=None)))*10\n",
    "I_rgbb[1,:,:]=np.double(np.divide(I2b,np.max(I2b,axis=None)))*10\n",
    "\n",
    "\n",
    "Irgbb=np.uint8(np.clip(I_rgbb*255,0,255))\n",
    "\n",
    "\n",
    "fg, ax=plt.subplots(1,2)\n",
    "ax[0].imshow(np.transpose(Irgb[:,1450:1550,1350:1450],(1,2,0)),vmin=0,vmax=1)\n",
    "ax[1].imshow(np.transpose(Irgbb[:,1450:1550,1350:1450],(1,2,0)),vmin=0,vmax=1)\n",
    "# ax[0].imshow(np.transpose(Irgb[:,:.:],(1,2,0)),vmin=0,vmax=1)\n",
    "# ax[1].imshow(np.transpose(Irgbb[:,:.:],(1,2,0)),vmin=0,vmax=1)\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84ad3ca8-8f90-4a5e-a92f-a2b25a504b8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#HYB FINAL LAYER-----------------develop\n",
    "# pthb='/mnt/d/allen_data/test_dataset_bc_726126/'\n",
    "\n",
    "# cycle_name='n2vhyb'\n",
    "# config_pth='/mnt/c/Users/goswami/Documents/MATLAB/barseq-matlab-allen-main (1)/barseq-matlab-allen-main/config/'\n",
    "# chshift_hyb_filename='chshift20x-hyb-20220218.mat'\n",
    "# chprofile_hyb_filename='chprofilehyb_allen.mat'\n",
    "# num_c=5\n",
    "# is_affine=0\n",
    "# cyclename='n2vhyb01'\n",
    "# strength=0.6\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a814de0-2b51-4511-8482-86b8acf20b0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PROCESS BC IMAGES\n",
    "pthb='/mnt/d/allen_data/test_dataset_bc_726126'\n",
    "\n",
    "process_bc_cycle(pthb,config_pth,chshift_filename,chprofile_filename,is_affine=0,num_initial_c=5,num_later_c=4,num_c=4,subsample_rate=4,do_coarse=0,resize_factor=2,block_size=128,radius=31,local=1,cyclename='regn2vbc',bcidx=4,geneidx=4,ch_bc=4,ch_bc_init=5,bcfname='n2vbcseq',genefname='n2vgene',strength=0.6,m=0)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4560c0e9-d4b3-45e8-8a18-41bb3c786c22",
   "metadata": {},
   "outputs": [],
   "source": [
    "# BARDENSR CALL\n",
    "t0=time.time()\n",
    "pth='/mnt/d/allen_data/processed/'\n",
    "bardensr_call(pth,config_pth,num_channels=4,codebook_name='codebookM1all.mat',fdrthresh=0.05,trim=160,cropf=0.4,noisefloor_ini=0.01,noisefloor_final=0.05)\n",
    "tn=time.time()\n",
    "print(f'Total time for {len(folders)} is {tn-t0} secs')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "370587b5-a20b-4a20-96e7-442c984345ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# IMPORT BARDENSR RESULTS ACROSS WHOLE DATASET\n",
    "pth='/mnt/d/allen_data/test_dataset_bc_726126/'\n",
    "fname='bardensrresult.csv'\n",
    "import_bardensr_results(pth,fname='bardensrresult.csv',prev_codebook_len=0,codebook_name='codebook.mat',id_optseq=0,codebook_opt_name='codebook.mat')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "415f7cc3-df17-4d5e-a2f0-71f6a59e8bcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# HYB BASE CALL------------------------------written in a way that finds max for all spots across channels\n",
    "from skimage.measure import label\n",
    "from skimage import color\n",
    "from skimage.morphology import extrema\n",
    "from skimage.util import img_as_float\n",
    "from skimage.exposure import rescale_intensity\n",
    "relaxed=1;\n",
    "no_deconv=1;\n",
    "filter_overlap=0;\n",
    "codebook_hyb_name='codebookhyb.mat'\n",
    "config_pth='/mnt/c/Users/goswami/Documents/MATLAB/barseq-matlab-allen-main (1)/barseq-matlab-allen-main/config/'\n",
    "pth='/mnt/d/allen_data/test_dataset_bc_726126/'\n",
    "cyclename='alignedn2vhyb'\n",
    "num_c=4\n",
    "thresh=[50,60,40,30]\n",
    "prominence=[50,10,40,10]\n",
    "all_genes_ch=2\n",
    "basecall_hyb_all_tiles(pth,cyclename,config_pth,codebook_hyb_name,thresh,prominence,num_c,all_genes_ch,no_deconv)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7a65a40-70ca-4607-bbb7-b39d586889d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# CHECK HYB BASE CALL\n",
    "fov=9\n",
    "codebook_hyb_name='codebookhyb.mat'\n",
    "config_pth='/mnt/c/Users/goswami/Documents/MATLAB/barseq-matlab-allen-main (1)/barseq-matlab-allen-main/config/'\n",
    "pth='/mnt/d/allen_data/test_dataset_bc_726126/'\n",
    "codebook_hyb_path=os.path.join(config_pth,codebook_hyb_name)\n",
    "codebook_hyb=scipy.io.loadmat(codebook_hyb_path)['codebookhyb']\n",
    "print(codebook_hyb[0])\n",
    "fg,ax=plt.subplots()\n",
    "for i in range(len(L['lroi_x'][fov])):\n",
    "    ax.scatter(L['lroi_y'][fov][i], L['lroi_x'][fov][i],s=3)#,c=col[i])\n",
    "    ax.invert_yaxis()\n",
    "plt.legend([codebook_hyb[0][0],codebook_hyb[1][0],codebook_hyb[2][0]])\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ff9c4e2-3626-4903-8e33-7b2c402860f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# STITCH--MIST\n",
    "sx,sy=stitch_mist_allpos(pth,pth_fiji,fname='n2vgeneseq01',stitching_channel=3,overlap=0.23,numcores=8,is_headless=True,grid_direction='UR',px=0.33,transform_rescale_factor=0.5,num_c=4)\n",
    "produce_tform_and_stitched_images(pth,sx,sy,intensity_scaling=3,tilesize=3200,transform_rescale_factor=0.5,display_additional_rescale=0.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39c9701d-83ae-4deb-84c2-45ff46d6ffb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# STITCH--ASHLAR---main stitching routine for ASHLAR is not implemented here, it is with John\n",
    "sx,sy=merge_ashlar_results(pth,transform_rescale_factor=0.5,num_c=4)\n",
    "produce_tform_and_stitched_images(pth,sx,sy,intensity_scaling=3,tilesize=3200,transform_rescale_factor=0.5,display_additional_rescale=0.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff69f1e2-768e-4fbe-8067-8a31c355e18b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# CHECK STITCHED OUTPUT-MIST\n",
    "pth='/mnt/d/allen_data/test_dataset_bc_726126/'\n",
    "stitch_output=os.path.join(pth,'processed','checkregistration_mist')\n",
    "geneseq_pos=tfl.imread(os.path.join(stitch_output,'Pos1_RGBalignedn2vgeneseq01.tif'))\n",
    "bcseq_pos=tfl.imread(os.path.join(stitch_output,'Pos1_RGBalignedregn2vbcseq01.tif'))\n",
    "hybseq_pos=tfl.imread(os.path.join(stitch_output,'Pos1_RGBalignedn2vhyb01.tif'))\n",
    "fg, ax=plt.subplots(1,3)\n",
    "ax[0].imshow(geneseq_pos)\n",
    "ax[1].imshow(bcseq_pos)\n",
    "ax[2].imshow(hybseq_pos)\n",
    "fg.set_figheight(10)\n",
    "fg.set_figwidth(10)\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd08a70e-0e5c-497a-8d57-d2ca74a0b6be",
   "metadata": {},
   "outputs": [],
   "source": [
    "# CHECK STITCHED OUTPUT-ASHLAR\n",
    "pth='/mnt/d/allen_data/test_dataset_bc_726126/'\n",
    "stitch_output=os.path.join(pth,'processed','checkregistration')\n",
    "geneseq_pos=tfl.imread(os.path.join(stitch_output,'Pos1_RGBalignedn2vgeneseq01.tif'))\n",
    "bcseq_pos=tfl.imread(os.path.join(stitch_output,'Pos1_RGBalignedregn2vbcseq01.tif'))\n",
    "hybseq_pos=tfl.imread(os.path.join(stitch_output,'Pos1_RGBalignedn2vhyb01.tif'))\n",
    "fg, ax=plt.subplots(1,3)\n",
    "ax[0].imshow(geneseq_pos)\n",
    "ax[1].imshow(bcseq_pos)\n",
    "ax[2].imshow(hybseq_pos)\n",
    "fg.set_figheight(10)\n",
    "fg.set_figwidth(10)\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc2c0ea2-31a3-4db1-b4dc-d164ac70ef0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#CELLPOSE\n",
    "outname='cell_mask_cyto3.tif'\n",
    "model_name='cyto3'\n",
    "prepare_cellpose_input(pth,nuc_ch=5,num_chyb=5,num_cgene=4,tilesize=3200)\n",
    "run_diff_env_scripts(\"cellpose_segmentation\",'cfg.yaml', pth='/mnt/d/allen_data/test_dataset_bc_726126/', diameter=40, outname=outname,model_name=model_name)\n",
    "import_cellpose_all_tiles(pth,dilation_radius=3,fname=outname)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33ca77fb-b459-4950-96c3-3fa7da298420",
   "metadata": {},
   "outputs": [],
   "source": [
    "# DATA AGGREGATION\n",
    "pth='/mnt/d/allen_data/test_dataset_bc_726126/'\n",
    "aggregate_cell_ids_gene_hyb(pth,is_bc)\n",
    "transform_downsized_coordinates(pth,is_bc))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "891f27fc-f09c-4369-8e59-d4123c9e3105",
   "metadata": {},
   "outputs": [],
   "source": [
    "# DATA ORGANIZATION\n",
    "organize_processed_data(pth,config_pth,is_optseq=0,hyb_codebook_name='codebookhyb.mat',starting_slice_idx=1,starting_fov_idx=1,dummy_cell_num=10000,tilesize=3200,\n",
    "                           fraction_border=0.07)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2777018-791a-49e9-8690-22adeaf09101",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# FILTER OVER;APPING NEURONS\n",
    "pth='/mnt/d/allen_data/test_dataset_bc_726126/'\n",
    "filter_overlapping_cells(pth,fname='alldata06092025.joblib',rescale_factor=0.5,px=0.33,box_half_width_um=5,search_radius_um=200)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dcf52dc5-a2a2-4b22-96bc-05af859da338",
   "metadata": {},
   "outputs": [],
   "source": [
    "# DISPLAY OVERLAP REMOVAL RESULTS\n",
    "data2=load(os.path.join(pth,'processed','filt_neurons.joblib'))\n",
    "data=load(os.path.join(pth,'processed','alldata06092025.joblib'))\n",
    "neurons=data['neurons']\n",
    "filt_neurons=data2['filt_neurons']\n",
    "id_to_keep=data2['removecells_all']\n",
    "fg, ax=plt.subplots(1,2)\n",
    "fg.set_figheight(5)\n",
    "fg.set_figwidth(10)\n",
    "plt.tight_layout()\n",
    "exp_mat=neurons['expmat'].todense() # I should do it in csr rather than dense--memory efficient\n",
    "total_exp_cell=np.asarray(np.sum(exp_mat,axis=1))\n",
    "id_to_keep\n",
    "ax[0].scatter(neurons['pos10x_x'],neurons['pos10x_y'],1,total_exp_cell,cmap=plt.cm.viridis_r)\n",
    "ax[1].scatter(filt_neurons['pos10x_x'],filt_neurons['pos10x_y'],1,total_exp_cell[id_to_keep],cmap=plt.cm.viridis_r)\n",
    "print(np.sum(id_to_keep))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5cc56b3-83d9-4925-9d50-2d9a6c6dccae",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
